















<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="last-modified" content="2022-04-26 15:35:11 +0000">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- meta "search-domain" used for google site search function google_search() -->
    <meta name="search-domain" value="">
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/lesson.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/syntax.css" />
    <link rel="license" href="#license-info" />
    <script type="text/javascript" src="../assets/js/tabs.js"></script>
    <link rel="stylesheet" href="../assets/css/tabs.css">

    

    



    <!-- Favicons for everyone -->
    <link rel="apple-touch-icon-precomposed" sizes="57x57" href="../assets/favicons/swc/apple-touch-icon-57x57.png" />
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="../assets/favicons/swc/apple-touch-icon-114x114.png" />
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="../assets/favicons/swc/apple-touch-icon-72x72.png" />
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="../assets/favicons/swc/apple-touch-icon-144x144.png" />
    <link rel="apple-touch-icon-precomposed" sizes="60x60" href="../assets/favicons/swc/apple-touch-icon-60x60.png" />
    <link rel="apple-touch-icon-precomposed" sizes="120x120" href="../assets/favicons/swc/apple-touch-icon-120x120.png" />
    <link rel="apple-touch-icon-precomposed" sizes="76x76" href="../assets/favicons/swc/apple-touch-icon-76x76.png" />
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../assets/favicons/swc/apple-touch-icon-152x152.png" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-196x196.png" sizes="196x196" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-96x96.png" sizes="96x96" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-16x16.png" sizes="16x16" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-128.png" sizes="128x128" />
    <meta name="application-name" content="Software Carpentry - Getting started with CWL"/>
    <meta name="msapplication-TileColor" content="#FFFFFF" />
    <meta name="msapplication-TileImage" content="../assets/favicons/swc/mstile-144x144.png" />
    <meta name="msapplication-square70x70logo" content="../assets/favicons/swc/mstile-70x70.png" />
    <meta name="msapplication-square150x150logo" content="../assets/favicons/swc/mstile-150x150.png" />
    <meta name="msapplication-wide310x150logo" content="../assets/favicons/swc/mstile-310x150.png" />
    <meta name="msapplication-square310x310logo" content="../assets/favicons/swc/mstile-310x310.png" />


    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
	<script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
	<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
	<![endif]-->

  <title>
  Getting started with CWL
  </title>

  </head>
  <body>

    


<div class="panel panel-default life-cycle">
  <div id="life-cycle" class="panel-body beta">
    This lesson is being piloted (Beta version)
  </div>
</div>




    <div class="container">
      
















  
  










<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      
      
      <a href="https://software-carpentry.org" class="pull-left">
        <img class="navbar-logo" src="../assets/img/swc-icon-blue.svg" alt="Software Carpentry logo" />
      </a>
      

      
      <a class="navbar-brand" href="../index.html">Home</a>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">

	
        <li><a href="../CODE_OF_CONDUCT.html">Code of Conduct</a></li>

        
	
        <li><a href="../setup.html">Setup</a></li>

        
        
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Episodes <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
            
            <li><a href="../01-introduction/index.html">Introduction</a></li>
            
            
            <li><a href="../02-workflow/index.html">Create a Workflow by Composing Tools</a></li>
            
            
            <li><a href="../03-running/index.html">Running and Debugging a Workflow</a></li>
            
            
            <li><a href="../04-commandlinetool/index.html">Writing a Tool Wrapper</a></li>
            
            
            <li><a href="../05-scatter/index.html">Analyzing Multiple Samples</a></li>
            
            
            <li><a href="../06-expressions/index.html">Dynamic Workflow Behavior</a></li>
            
            
            <li><a href="../07-resources/index.html"> Resources for further learning</a></li>
            
            
            <li><a href="../08-supplement-docker/index.html">Supplement: Creating Docker Images for Workflows</a></li>
            
	    <li role="separator" class="divider"></li>
            <li><a href="../aio/index.html">All in one page (Beta)</a></li>
          </ul>
        </li>
        
	

	
	
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Extras <span class="caret"></span></a>
          <ul class="dropdown-menu">
            <li><a href="../reference.html">Reference</a></li>
            
            
            <li><a href="../about/index.html">About</a></li>
            
            
            <li><a href="../discuss/index.html">Discussion</a></li>
            
            
            <li><a href="../figures/index.html">Figures</a></li>
            
            
            <li><a href="../guide/index.html">Instructor Notes</a></li>
            
          </ul>
        </li>
	

	
        <li><a href="../LICENSE.html">License</a></li>
	
	<li><a href="/edit//aio.md" data-checker-ignore>Improve this page <span class="glyphicon glyphicon-pencil" aria-hidden="true"></span></a></li>
	
      </ul>
      <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form>
    </div>
  </div>
</nav>
















<h1 class="maintitle"><a href="../index.html">Getting started with CWL</a></h1>



<article>

<h1 id="introduction" class="maintitle">Introduction</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 10 min
      <br />
      <strong>Exercises:</strong> 0 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>What is CWL?</p>
</li>
	
	<li><p>What is the goal of this training?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Gain a high level understanding of the example analysis.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<p>This lesson has migrated to https://doc.arvados.org/rnaseq-cwl-training/01-introduction/index.html</p>

<h1 id="introduction-to-common-worklow-language">Introduction to Common Worklow Language</h1>

<p>The Common Workflow Language (CWL) is an open standard for describing
automated, batch data analysis workflows.  Unlike many programming
languages, CWL is a declarative language.  This means it describes
<em>what</em> should happen, but not <em>how</em> it should happen.  This enables
workflows written in CWL to be portable and scalable across a variety
of software and hardware environments, from workstations to cluster,
cloud, and high performance computing (HPC) environments.  As a
standard with multiple implementations, CWL is particularly well
suited for research collaboration, publishing, and high-throughput
production data analysis.</p>

<h1 id="introduction-to-the-example-analysis">Introduction to the example analysis</h1>

<p>This training uses a bioinformatics RNA-seq analysis as a motivating
example.  However, specific knowledge of the biology of RNA-seq is
<em>not</em> required for these lessons.  For those unfamiliar with RNA-seq,
it is the process of sequencing RNA present in a biological sample.
From the sequence reads, we want to measure the relative numbers of
different RNA molecules appearing in the sample that were produced by
particular genes.  This analysis is called “differential gene
expression”.</p>

<p>The entire process looks like this:</p>

<p><img src="../assets/img/RNAseqWorkflow.png" alt="" height="400px" /></p>

<p>For this training, we are only concerned with the middle analytical
steps (skipping adapter trimming).</p>

<ul>
  <li>Quality control (FASTQC)</li>
  <li>Alignment (mapping)</li>
  <li>Counting reads associated with genes</li>
</ul>

<p>In this training, we do not develop the analysis from first
principals, instead we we will be starting from an analysis already
written as a shell script, which will be presented in lesson 2.</p>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>Common Workflow Language is a standard for describing data analysis workflows</p>
</li>
    
    <li><p>We will use an bioinformatics RNA-seq analysis as an example workflow, but does not require in-depth knowledge of biology.</p>
</li>
    
    <li><p>After completing this training, you should be able to begin writing workflows for your own analysis, and know where to learn more.</p>
</li>
    
  </ul>
</blockquote>

<hr />

<h1 id="create-a-workflow-by-composing-tools" class="maintitle">Create a Workflow by Composing Tools</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 30 min
      <br />
      <strong>Exercises:</strong> 10 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>What is the syntax of CWL?</p>
</li>
	
	<li><p>What are the key components of a workflow?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Write a workflow based on the source shell script, making use of existing tool wrappers.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<p>This lesson has migrated to https://doc.arvados.org/rnaseq-cwl-training/02-workflow/index.html</p>

<h1 id="source-shell-script">Source shell script</h1>

<p>In this lesson, we will develop an initial workflow inspired by the
following shell script.</p>

<p>rnaseq_analysis_on_input_file.sh</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">#!/bin/bash</span>

<span class="c"># Based on</span>
<span class="c"># https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/07_automating_workflow.html</span>
<span class="c">#</span>

<span class="c"># This script takes a fastq file of RNA-Seq data, runs FastQC and outputs a counts file for it.</span>
<span class="c"># USAGE: sh rnaseq_analysis_on_input_file.sh &lt;name of fastq file&gt;</span>

<span class="nb">set</span> <span class="nt">-e</span>

<span class="c"># initialize a variable with an intuitive name to store the name of the input fastq file</span>
<span class="nv">fq</span><span class="o">=</span><span class="nv">$1</span>

<span class="c"># grab base of filename for naming outputs</span>
<span class="nv">base</span><span class="o">=</span><span class="sb">`</span><span class="nb">basename</span> <span class="nv">$fq</span> .subset.fq<span class="sb">`</span>
<span class="nb">echo</span> <span class="s2">"Sample name is </span><span class="nv">$base</span><span class="s2">"</span>

<span class="c"># specify the number of cores to use</span>
<span class="nv">cores</span><span class="o">=</span>4

<span class="c"># directory with genome reference FASTA and index files + name of the gene annotation file</span>
<span class="nv">genome</span><span class="o">=</span>rnaseq/reference_data
<span class="nv">gtf</span><span class="o">=</span>rnaseq/reference_data/chr1-hg19_genes.gtf

<span class="c"># make all of the output directories</span>
<span class="c"># The -p option means mkdir will create the whole path if it</span>
<span class="c"># does not exist and refrain from complaining if it does exist</span>
<span class="nb">mkdir</span> <span class="nt">-p</span> rnaseq/results/fastqc
<span class="nb">mkdir</span> <span class="nt">-p</span> rnaseq/results/STAR
<span class="nb">mkdir</span> <span class="nt">-p</span> rnaseq/results/counts

<span class="c"># set up output filenames and locations</span>
<span class="nv">fastqc_out</span><span class="o">=</span>rnaseq/results/fastqc
<span class="nv">align_out</span><span class="o">=</span>rnaseq/results/STAR/<span class="k">${</span><span class="nv">base</span><span class="k">}</span>_
<span class="nv">counts_input_bam</span><span class="o">=</span>rnaseq/results/STAR/<span class="k">${</span><span class="nv">base</span><span class="k">}</span>_Aligned.sortedByCoord.out.bam
<span class="nv">counts</span><span class="o">=</span>rnaseq/results/counts/<span class="k">${</span><span class="nv">base</span><span class="k">}</span>_featurecounts.txt

<span class="nb">echo</span> <span class="s2">"Processing file </span><span class="nv">$fq</span><span class="s2">"</span>

<span class="c"># Run FastQC and move output to the appropriate folder</span>
fastqc <span class="nv">$fq</span>

<span class="c"># Run STAR</span>
STAR <span class="nt">--runThreadN</span> <span class="nv">$cores</span> <span class="nt">--genomeDir</span> <span class="nv">$genome</span> <span class="nt">--readFilesIn</span> <span class="nv">$fq</span> <span class="nt">--outSAMtype</span> BAM SortedByCoordinate <span class="nt">--outSAMunmapped</span> Within

<span class="c"># Create BAM index</span>
samtools index <span class="nv">$counts_input_bam</span>

<span class="c"># Count mapped reads</span>
featureCounts <span class="nt">-T</span> <span class="nv">$cores</span> <span class="nt">-s</span> 2 <span class="nt">-a</span> <span class="nv">$gtf</span> <span class="nt">-o</span> <span class="nv">$counts</span> <span class="nv">$counts_input_bam</span>
</code></pre></div></div>

<h1 id="cwl-syntax">CWL Syntax</h1>

<p>CWL documents are written using a format called “YAML”.  Here is a crash-course in YAML:</p>

<p>Data fields are written with the name, followed by a colon <code class="language-plaintext highlighter-rouge">:</code>, a space,
and then the value.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">fieldName</span><span class="pi">:</span> <span class="s">value</span>
</code></pre></div></div>

<p>The value is the remaining text to the end of the line.</p>

<p>Special characters in YAML include <code class="language-plaintext highlighter-rouge">:</code>, <code class="language-plaintext highlighter-rouge">{</code>, <code class="language-plaintext highlighter-rouge">}</code> <code class="language-plaintext highlighter-rouge">[</code>, <code class="language-plaintext highlighter-rouge">]</code>, <code class="language-plaintext highlighter-rouge">#</code>, <code class="language-plaintext highlighter-rouge">!</code>
and <code class="language-plaintext highlighter-rouge">%</code>.  If your text begins with any of these characters, you must
surround the string in single or double quotes.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">fieldName</span><span class="pi">:</span> <span class="s2">"</span><span class="s">#quoted-value"</span>
</code></pre></div></div>

<p>You can write multi-line text by putting <code class="language-plaintext highlighter-rouge">|-</code> and writing an indented
block.  The leading whitespace will be removed from the actual value.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">fieldName</span><span class="pi">:</span> <span class="pi">|-</span>
  <span class="s">This is a multi-</span>
  <span class="s">line string.</span>
  <span class="s">Horray!</span>
</code></pre></div></div>

<p>Nested sections are indented:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">section1</span><span class="pi">:</span>
  <span class="na">field1</span><span class="pi">:</span> <span class="s">value1</span>
  <span class="na">field2</span><span class="pi">:</span> <span class="s">value2</span>
</code></pre></div></div>

<p>Nested sections can <em>also</em> be wrapped in curly brackets.  In this case, fields must be comma-separated.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">section1</span><span class="pi">:</span> <span class="pi">{</span><span class="nv">field1</span><span class="pi">:</span> <span class="nv">value1</span><span class="pi">,</span> <span class="nv">field2</span><span class="pi">,</span> <span class="nv">value2</span><span class="pi">}</span>
</code></pre></div></div>

<p>When each item is on its own line starting with a dash <code class="language-plaintext highlighter-rouge">-</code>, it is a list.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">section2</span><span class="pi">:</span>
  <span class="pi">-</span> <span class="s">value1</span>
  <span class="pi">-</span> <span class="s">value2</span>
</code></pre></div></div>

<p>List can <em>also</em> be wrapped in square brackets.  In this case, values must be comma-separated.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">section2</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">value1</span><span class="pi">,</span> <span class="nv">value2</span><span class="pi">]</span>
</code></pre></div></div>

<p>Comments start with <code class="language-plaintext highlighter-rouge">#</code>.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># This is a comment about field3</span>
<span class="na">field3</span><span class="pi">:</span> <span class="s">stuff</span>

<span class="na">field4</span><span class="pi">:</span> <span class="s">stuff</span> <span class="c1"># This is a comment about field4</span>
</code></pre></div></div>

<p>Finally, YAML is a superset of JSON.  Valid JSON is also valid YAML,
so you may sometimes see JSON format being used instead of YAML format
for CWL documents.</p>

<h1 id="workflow-header">Workflow header</h1>

<p>Create a new file “main.cwl”</p>

<p>Let’s start with the header.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">cwlVersion</span><span class="pi">:</span> <span class="s">v1.2</span>
<span class="na">class</span><span class="pi">:</span> <span class="s">Workflow</span>
<span class="na">label</span><span class="pi">:</span> <span class="s">RNAseq CWL practice workflow</span>
</code></pre></div></div>

<ul>
  <li>cwlVersion - Every file must include this.  It declares the version of CWL in use.</li>
  <li>class - This is the type of CWL document.  We will see other types in future lessons.</li>
  <li>label - Optional title of your workflow.</li>
</ul>

<h1 id="workflow-inputs">Workflow Inputs</h1>

<p>The purpose of a workflow is to consume some input parameters, run a
series of steps, and produce output values.</p>

<p>For this analysis, the input parameters are the fastq file and the reference data required by STAR.</p>

<p>In the source shell script, the following variables are declared:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># initialize a variable with an intuitive name to store the name of the input fastq file</span>
<span class="nv">fq</span><span class="o">=</span><span class="nv">$1</span>

<span class="c"># directory with genome reference FASTA and index files + name of the gene annotation file</span>
<span class="nv">genome</span><span class="o">=</span>rnaseq/reference_data
<span class="nv">gtf</span><span class="o">=</span>rnaseq/reference_data/chr1-hg19_genes.gtf
</code></pre></div></div>

<p>In CWL, we will declare these variables in the <code class="language-plaintext highlighter-rouge">inputs</code> section.</p>

<p>The inputs section lists each input parameter and its type.  Valid
types include <code class="language-plaintext highlighter-rouge">File</code>, <code class="language-plaintext highlighter-rouge">Directory</code>, <code class="language-plaintext highlighter-rouge">string</code>, <code class="language-plaintext highlighter-rouge">boolean</code>, <code class="language-plaintext highlighter-rouge">int</code>, and
<code class="language-plaintext highlighter-rouge">float</code>.</p>

<p>In this case, the fastq and gene annotation file are individual files.  The STAR index is a directory.  We can describe these inputs in CWL like this:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">inputs</span><span class="pi">:</span>
  <span class="na">fq</span><span class="pi">:</span> <span class="s">File</span>
  <span class="na">genome</span><span class="pi">:</span> <span class="s">Directory</span>
  <span class="na">gtf</span><span class="pi">:</span> <span class="s">File</span>
</code></pre></div></div>

<h1 id="workflow-steps">Workflow Steps</h1>

<p>A workflow consists of one or more steps.  This is the <code class="language-plaintext highlighter-rouge">steps</code> section.</p>

<p>Now we need to describe the first step of the workflow.  In the source
script, the first step is to run <code class="language-plaintext highlighter-rouge">fastqc</code>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Run FastQC and move output to the appropriate folder</span>
fastqc <span class="nv">$fq</span>
</code></pre></div></div>

<p>A workflow step consists of the name of the step, the tool to <code class="language-plaintext highlighter-rouge">run</code>,
the input parameters to be passed to the tool in <code class="language-plaintext highlighter-rouge">in</code>, and the output
parameters expected from the tool in <code class="language-plaintext highlighter-rouge">out</code>.</p>

<p>The value of <code class="language-plaintext highlighter-rouge">run</code> references the tool file.  The tool file describes
how to run the tool (we will discuss how to write tool files in lesson
4).  If we look in <code class="language-plaintext highlighter-rouge">bio-cwl-tools</code> (which you should have imported
when setting up a practice repository in the initial setup
instructions) we find <code class="language-plaintext highlighter-rouge">bio-cwl-tools/fastqc/fastqc_2.cwl</code>.</p>

<p>Next, the <code class="language-plaintext highlighter-rouge">in</code> block is mapping of input parameters to the tool and
the workflow parameters that will be assigned to those inputs.  We
need to know what input parameters the tool accepts.</p>

<p>Let’s open up the tool file and take a look:</p>

<p>Find the <code class="language-plaintext highlighter-rouge">inputs</code> section of <code class="language-plaintext highlighter-rouge">bio-cwl-tools/fastqc/fastqc_2.cwl</code>:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">inputs</span><span class="pi">:</span>

  <span class="na">reads_file</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span>
      <span class="pi">-</span> <span class="s">File</span>
    <span class="na">inputBinding</span><span class="pi">:</span>
      <span class="na">position</span><span class="pi">:</span> <span class="m">50</span>
    <span class="na">doc</span><span class="pi">:</span> <span class="pi">|</span>
      <span class="s">Input bam,sam,bam_mapped,sam_mapped or fastq file</span>
</code></pre></div></div>

<p>Now we know we need to provide an input parameter called <code class="language-plaintext highlighter-rouge">reads_file</code>.</p>

<p>Next, the <code class="language-plaintext highlighter-rouge">out</code> section is a list of output parameters from the tool
that will be used later in the workflow, or as workflow output.  We
need to know what output parameters the tool produces.  Find the
<code class="language-plaintext highlighter-rouge">outputs</code> section of <code class="language-plaintext highlighter-rouge">bio-cwl-tools/fastqc/fastqc_2.cwl</code>:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">outputs</span><span class="pi">:</span>

  <span class="na">zipped_file</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span>
      <span class="pi">-</span> <span class="s">File</span>
    <span class="na">outputBinding</span><span class="pi">:</span>
      <span class="na">glob</span><span class="pi">:</span> <span class="s1">'</span><span class="s">*.zip'</span>
  <span class="na">html_file</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span>
      <span class="pi">-</span> <span class="s">File</span>
    <span class="na">outputBinding</span><span class="pi">:</span>
      <span class="na">glob</span><span class="pi">:</span> <span class="s1">'</span><span class="s">*.html'</span>
  <span class="na">summary_file</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span>
      <span class="pi">-</span> <span class="s">File</span>
    <span class="na">outputBinding</span><span class="pi">:</span>
      <span class="na">glob</span><span class="pi">:</span> <span class="pi">|</span>
        <span class="s">${</span>
          <span class="s">return "*/summary.txt";</span>
        <span class="s">}</span>
</code></pre></div></div>

<p>Now we know to expect an output parameter called <code class="language-plaintext highlighter-rouge">html_file</code>.</p>

<p>Putting this all together, the <code class="language-plaintext highlighter-rouge">fastq</code> step consists of a <code class="language-plaintext highlighter-rouge">run</code>, <code class="language-plaintext highlighter-rouge">in</code>
and <code class="language-plaintext highlighter-rouge">out</code> subsections, and looks like this:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">steps</span><span class="pi">:</span>
  <span class="na">fastqc</span><span class="pi">:</span>
    <span class="na">run</span><span class="pi">:</span> <span class="s">bio-cwl-tools/fastqc/fastqc_2.cwl</span>
    <span class="na">in</span><span class="pi">:</span>
      <span class="na">reads_file</span><span class="pi">:</span> <span class="s">fq</span>
    <span class="na">out</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">html_file</span><span class="pi">]</span>
</code></pre></div></div>

<h1 id="running-alignment-with-star">Running alignment with STAR</h1>

<p>The next step is to run the STAR aligner.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Run STAR</span>
STAR <span class="nt">--runThreadN</span> <span class="nv">$cores</span> <span class="nt">--genomeDir</span> <span class="nv">$genome</span> <span class="nt">--readFilesIn</span> <span class="nv">$fq</span> <span class="nt">--outSAMtype</span> BAM SortedByCoordinate <span class="nt">--outSAMunmapped</span> Within
</code></pre></div></div>

<p>We will go through the same process as the first section.  We find
there is <code class="language-plaintext highlighter-rouge">bio-cwl-tools/STAR/STAR-Align.cwl</code>.  We will open the file
and look at the <code class="language-plaintext highlighter-rouge">inputs</code> section to determine what input parameters
correspond to the command line parmeters from our source script.</p>

<blockquote class="challenge">
  <h2 id="exercise">Exercise</h2>

  <p>Look at <code class="language-plaintext highlighter-rouge">STAR-Align.cwl</code> and identify the input parameters that
correspond to the command line arguments used in the source script:
<code class="language-plaintext highlighter-rouge">--runThreadN</code>, <code class="language-plaintext highlighter-rouge">--genomeDir</code>, <code class="language-plaintext highlighter-rouge">--outSAMtype</code>, and
<code class="language-plaintext highlighter-rouge">--outSAMunmapped</code>.  Also identify the name of the output parameter.</p>

  <blockquote class="solution">
    <h2 id="solution">Solution</h2>

    <ul>
      <li><code class="language-plaintext highlighter-rouge">--runThreadN</code> → <code class="language-plaintext highlighter-rouge">RunThreadN</code></li>
      <li><code class="language-plaintext highlighter-rouge">--genomeDir</code> → <code class="language-plaintext highlighter-rouge">GenomeDir</code></li>
      <li><code class="language-plaintext highlighter-rouge">--readFilesIn</code> → <code class="language-plaintext highlighter-rouge">ForwardReads</code></li>
      <li><code class="language-plaintext highlighter-rouge">--outSAMtype</code> → <code class="language-plaintext highlighter-rouge">OutSAMtype</code>, <code class="language-plaintext highlighter-rouge">SortedByCoordinate</code></li>
      <li><code class="language-plaintext highlighter-rouge">--outSAMunmapped</code> → <code class="language-plaintext highlighter-rouge">OutSAMunmapped</code></li>
    </ul>

    <p>output parameter name: <code class="language-plaintext highlighter-rouge">alignment</code></p>
  </blockquote>
</blockquote>

<blockquote class="callout">
  <h2 id="command-line-flags">Command line flags</h2>

  <p>Command line flags generally appear appear in either the <code class="language-plaintext highlighter-rouge">arguments</code>
field, or the <code class="language-plaintext highlighter-rouge">prefix</code> field of the <code class="language-plaintext highlighter-rouge">inputBinding</code> section of an
input parameter declaration.  For example, this section of
<code class="language-plaintext highlighter-rouge">STAR-Align.cwl</code> tells us that the <code class="language-plaintext highlighter-rouge">GenomeDir</code> input parameter
corresponds to the <code class="language-plaintext highlighter-rouge">--genomeDir</code> command line parameter.</p>

  <div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="na">GenomeDir</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">Directory</span>
    <span class="na">inputBinding</span><span class="pi">:</span>
      <span class="na">prefix</span><span class="pi">:</span> <span class="s2">"</span><span class="s">--genomeDir"</span>
</code></pre></div>  </div>
</blockquote>

<blockquote class="callout">
  <h2 id="default-values">Default values</h2>

  <p>Sometimes we want to provide input values to a step without making
them as workflow-level inputs.  We can do this with <code class="language-plaintext highlighter-rouge">{default: N}</code>.
For example:</p>

  <div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code>   <span class="na">in</span><span class="pi">:</span>
     <span class="na">RunThreadN</span><span class="pi">:</span> <span class="pi">{</span><span class="nv">default</span><span class="pi">:</span> <span class="nv">4</span><span class="pi">}</span>
</code></pre></div>  </div>
</blockquote>

<blockquote class="challenge">
  <h2 id="exercise-1">Exercise</h2>

  <p>Using the input and output parameters identified in the previous
exercise, write the <code class="language-plaintext highlighter-rouge">run</code>, <code class="language-plaintext highlighter-rouge">in</code> and <code class="language-plaintext highlighter-rouge">out</code> sections of the STAR step.</p>

  <blockquote class="solution">
    <h2 id="solution-1">Solution</h2>

    <div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="na">STAR</span><span class="pi">:</span>
    <span class="na">run</span><span class="pi">:</span> <span class="s">bio-cwl-tools/STAR/STAR-Align.cwl</span>
    <span class="na">in</span><span class="pi">:</span>
      <span class="na">RunThreadN</span><span class="pi">:</span> <span class="pi">{</span><span class="nv">default</span><span class="pi">:</span> <span class="nv">4</span><span class="pi">}</span>
      <span class="na">GenomeDir</span><span class="pi">:</span> <span class="s">genome</span>
      <span class="na">ForwardReads</span><span class="pi">:</span> <span class="s">fq</span>
      <span class="na">OutSAMtype</span><span class="pi">:</span> <span class="pi">{</span><span class="nv">default</span><span class="pi">:</span> <span class="nv">BAM</span><span class="pi">}</span>
      <span class="na">SortedByCoordinate</span><span class="pi">:</span> <span class="pi">{</span><span class="nv">default</span><span class="pi">:</span> <span class="nv">true</span><span class="pi">}</span>
      <span class="na">OutSAMunmapped</span><span class="pi">:</span> <span class="pi">{</span><span class="nv">default</span><span class="pi">:</span> <span class="nv">Within</span><span class="pi">}</span>
    <span class="na">out</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">alignment</span><span class="pi">]</span>
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<h1 id="running-samtools">Running samtools</h1>

<p>The third step is to generate an index for the aligned BAM.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Create BAM index</span>
samtools index <span class="nv">$counts_input_bam</span>
</code></pre></div></div>

<p>For this step, we need to use the output of a previous step as input
to this step.  We refer the output of a step by with name of the step
(STAR), a slash, and the name of the output parameter (alignment), e.g. <code class="language-plaintext highlighter-rouge">STAR/alignment</code></p>

<p>This creates a dependency between steps.  This means the <code class="language-plaintext highlighter-rouge">samtools</code>
step will not run until the <code class="language-plaintext highlighter-rouge">STAR</code> step has completed successfully.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="na">samtools</span><span class="pi">:</span>
    <span class="na">run</span><span class="pi">:</span> <span class="s">bio-cwl-tools/samtools/samtools_index.cwl</span>
    <span class="na">in</span><span class="pi">:</span>
      <span class="na">bam_sorted</span><span class="pi">:</span> <span class="s">STAR/alignment</span>
    <span class="na">out</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">bam_sorted_indexed</span><span class="pi">]</span>
</code></pre></div></div>

<h1 id="featurecounts">featureCounts</h1>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Count mapped reads</span>
featureCounts <span class="nt">-T</span> <span class="nv">$cores</span> <span class="nt">-s</span> 2 <span class="nt">-a</span> <span class="nv">$gtf</span> <span class="nt">-o</span> <span class="nv">$counts</span> <span class="nv">$counts_input_bam</span>
</code></pre></div></div>

<p>As of this writing, the <code class="language-plaintext highlighter-rouge">subread</code> package that provides
<code class="language-plaintext highlighter-rouge">featureCounts</code> is not available in <code class="language-plaintext highlighter-rouge">bio-cwl-tools</code> (and if it has been
added since then, let’s pretend that it isn’t there.)  We will go over
how to write a CWL wrapper for a command line tool in lesson 4.  For
now, we will leave off the final step.</p>

<h1 id="workflow-outputs">Workflow Outputs</h1>

<p>The last thing to do is declare the workflow outputs in the <code class="language-plaintext highlighter-rouge">outputs</code> section.</p>

<p>For each output, we need to declare the type of output, and what
parameter has the output value.</p>

<p>Output types are the same as input types, valid types include <code class="language-plaintext highlighter-rouge">File</code>,
<code class="language-plaintext highlighter-rouge">Directory</code>, <code class="language-plaintext highlighter-rouge">string</code>, <code class="language-plaintext highlighter-rouge">boolean</code>, <code class="language-plaintext highlighter-rouge">int</code>, and <code class="language-plaintext highlighter-rouge">float</code>.</p>

<p>The <code class="language-plaintext highlighter-rouge">outputSource</code> field refers the a step output in the same way that
the <code class="language-plaintext highlighter-rouge">in</code> block does, the name of the step, a slash, and the name of
the output parameter.</p>

<p>For our final outputs, we want the results from fastqc and the
aligned, sorted and indexed BAM file.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">outputs</span><span class="pi">:</span>
  <span class="na">qc_html</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">outputSource</span><span class="pi">:</span> <span class="s">fastqc/html_file</span>
  <span class="na">bam_sorted_indexed</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">outputSource</span><span class="pi">:</span> <span class="s">samtools/bam_sorted_indexed</span>
</code></pre></div></div>

<blockquote class="solution">
  <h2 id="episode-solution">Episode solution</h2>
  <ul>
    <li><a href="../assets/answers/ep2/main.cwl">main.cwl</a></li>
  </ul>
</blockquote>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>CWL documents are written using a syntax called YAML.</p>
</li>
    
    <li><p>The key components of the workflow are: the header, the inputs, the steps, and the outputs.</p>
</li>
    
  </ul>
</blockquote>

<hr />

<h1 id="running-and-debugging-a-workflow" class="maintitle">Running and Debugging a Workflow</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 15 min
      <br />
      <strong>Exercises:</strong> 20 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>How do I provide input to run a workflow?</p>
</li>
	
	<li><p>What should I do if the workflow fails?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Write an input parameter file.</p>
</li>
	
	<li><p>Execute the workflow.</p>
</li>
	
	<li><p>Diagnose workflow errors.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<p>This lesson has migrated to https://doc.arvados.org/rnaseq-cwl-training/03-running/index.html</p>

<h1 id="the-input-parameter-file">The input parameter file</h1>

<p>CWL input values are provided in the form of a YAML or JSON file.
create a called file</p>

<p>This file gives the values for parameters declared in the <code class="language-plaintext highlighter-rouge">inputs</code>
section of our workflow.  Our workflow takes <code class="language-plaintext highlighter-rouge">fq</code>, <code class="language-plaintext highlighter-rouge">genome</code> and <code class="language-plaintext highlighter-rouge">gtf</code>
as input parameters.</p>

<p>When setting inputs, Files and Directories are given as an object with
<code class="language-plaintext highlighter-rouge">class: File</code> or <code class="language-plaintext highlighter-rouge">class: Directory</code>.  This distinguishes them from
plain strings that may or may not be file paths.</p>

<p>Note: if you don’t have example sequence data or the STAR index files, see <a href="/setup.html">setup</a>.</p>

<div class="tabbed">
  <ul class="tab">
      <li><a href="#section-generic-input">generic</a></li>
      <li><a href="#section-arvados-input">arvados</a></li>
  </ul>

  <section id="section-generic-input">
<p>main-input.yaml</p>
<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">fq</span><span class="pi">:</span>
  <span class="na">class</span><span class="pi">:</span> <span class="s">File</span>
  <span class="na">location</span><span class="pi">:</span> <span class="s">rnaseq/raw_fastq/Mov10_oe_1.subset.fq</span>
  <span class="na">format</span><span class="pi">:</span> <span class="s">http://edamontology.org/format_1930</span>
<span class="na">genome</span><span class="pi">:</span>
  <span class="na">class</span><span class="pi">:</span> <span class="s">Directory</span>
  <span class="na">location</span><span class="pi">:</span> <span class="s">hg19-chr1-STAR-index</span>
<span class="na">gtf</span><span class="pi">:</span>
  <span class="na">class</span><span class="pi">:</span> <span class="s">File</span>
  <span class="na">location</span><span class="pi">:</span> <span class="s">rnaseq/reference_data/chr1-hg19_genes.gtf</span>
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="running-the-workflow">Running the workflow</h2>

  <p>Type this into the terminal:</p>

  <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cwl-runner main.cwl main-input.yaml
</code></pre></div>  </div>

  <p>This may take a few minutes to run, and will print some amount of
logging.  The logging you see, how access other logs, and how to
track workflow progress will depend on your CWL runner platform.</p>
</blockquote>
</section>
  <section id="section-arvados-input">
<p>main-input.yaml</p>
<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">fq</span><span class="pi">:</span>
  <span class="na">class</span><span class="pi">:</span> <span class="s">File</span>
  <span class="na">location</span><span class="pi">:</span> <span class="s">keep:9178fe1b80a08a422dbe02adfd439764+925/raw_fastq/Mov10_oe_1.subset.fq</span>
  <span class="na">format</span><span class="pi">:</span> <span class="s">http://edamontology.org/format_1930</span>
<span class="na">genome</span><span class="pi">:</span>
  <span class="na">class</span><span class="pi">:</span> <span class="s">Directory</span>
  <span class="na">location</span><span class="pi">:</span> <span class="s">keep:02a12ce9e2707610991bd29d38796b57+2912</span>
<span class="na">gtf</span><span class="pi">:</span>
  <span class="na">class</span><span class="pi">:</span> <span class="s">File</span>
  <span class="na">location</span><span class="pi">:</span> <span class="s">keep:9178fe1b80a08a422dbe02adfd439764+925/reference_data/chr1-hg19_genes.gtf</span>
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="running-the-workflow">Running the workflow</h2>

  <p>If you are using VSCode with Arvados tasks, select <code class="language-plaintext highlighter-rouge">main.cwl</code> and
then use the <code class="language-plaintext highlighter-rouge">Run CWL Workflow on Arvados</code> task.</p>

</blockquote>
</section>
</div>

<h1 id="debugging-the-workflow">Debugging the workflow</h1>

<p>Depending on whether and how your workflow platform enforces memory
limits, your workflow may fail.  Let’s talk about what to do when a
workflow fails.</p>

<p>A workflow can fail for many reasons: some possible reasons include
bad input, bugs in the code, or running out memory.  In our example,
the STAR workflow may fail with an out of memory error.</p>

<p>To help diagnose these errors, the workflow runner produces logs that
record what happened, either in the terminal or the web interface.</p>

<p>Some errors you might see in the logs that would indicate an out of
memory condition:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>EXITING: fatal error trying to allocate genome arrays, exception thrown: std::bad_alloc
Possible cause 1: not enough RAM. Check if you have enough RAM 5711762337 bytes
Possible cause 2: not enough virtual memory allowed with ulimit. SOLUTION: run ulimit -v 5711762337
</code></pre></div></div>

<p>or</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Container exited with code: 137
</code></pre></div></div>

<p>(Exit code 137 most commonly occurs when a container goes “out of memory” and is terminated by the operating system).</p>

<p>If this happens, you will need to request more RAM.</p>

<h1 id="setting-runtime-ram-requirements">Setting runtime RAM requirements</h1>

<p>By default, a step is allocated 256 MB of RAM.  From the STAR error message:</p>

<blockquote>
  <p>Check if you have enough RAM 5711762337 bytes</p>
</blockquote>

<p>We can see that STAR requires quite a bit more RAM than 256 MB.  To
request more RAM, add a “requirements” section with
“ResourceRequirement” to the “STAR” step:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="na">STAR</span><span class="pi">:</span>
    <span class="na">requirements</span><span class="pi">:</span>
      <span class="na">ResourceRequirement</span><span class="pi">:</span>
        <span class="na">ramMin</span><span class="pi">:</span> <span class="m">9000</span>
    <span class="na">run</span><span class="pi">:</span> <span class="s">bio-cwl-tools/STAR/STAR-Align.cwl</span>
<span class="err">	</span><span class="s">...</span>
</code></pre></div></div>

<p>Resource requirements you can set include:</p>

<ul>
  <li>coresMin: CPU cores</li>
  <li>ramMin: RAM (in megabytes)</li>
  <li>tmpdirMin: temporary directory available space</li>
  <li>outdirMin: output directory available space</li>
</ul>

<blockquote class="challenge">
  <h2 id="running-the-workflow">Running the workflow</h2>

  <p>Now that you’ve fixed the workflow, run it again.</p>

</blockquote>

<blockquote class="solution">
  <h2 id="episode-solution">Episode solution</h2>
  <ul>
    <li><a href="../assets/answers/ep3/main.cwl">main.cwl</a></li>
  </ul>
</blockquote>

<h1 id="workflow-results">Workflow results</h1>

<p>The CWL runner will print a results JSON object to standard output.  It will look something like this (it may include additional fields).</p>

<div class="tabbed">
  <ul class="tab">
      <li><a href="#section-generic-output">generic</a></li>
      <li><a href="#section-arvados-output">arvados</a></li>
  </ul>

  <section id="section-generic-output">
<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="pi">{</span>
    <span class="s2">"</span><span class="s">bam_sorted_indexed"</span><span class="pi">:</span> <span class="pi">{</span>
        <span class="s2">"</span><span class="s">location"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">file:///home/username/rnaseq-cwl-training-exercises/Aligned.sortedByCoord.out.bam"</span><span class="pi">,</span>
        <span class="s2">"</span><span class="s">basename"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">Aligned.sortedByCoord.out.bam"</span><span class="pi">,</span>
        <span class="s2">"</span><span class="s">class"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">File"</span><span class="pi">,</span>
        <span class="s2">"</span><span class="s">size"</span><span class="pi">:</span> <span class="nv">25370707</span><span class="pi">,</span>
        <span class="s2">"</span><span class="s">secondaryFiles"</span><span class="pi">:</span> <span class="pi">[</span>
            <span class="pi">{</span>
                <span class="s2">"</span><span class="s">basename"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">Aligned.sortedByCoord.out.bam.bai"</span><span class="pi">,</span>
                <span class="s2">"</span><span class="s">location"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">file:///home/username/rnaseq-cwl-training-exercises/Aligned.sortedByCoord.out.bam.bai"</span><span class="pi">,</span>
                <span class="s2">"</span><span class="s">class"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">File"</span><span class="pi">,</span>
                <span class="s2">"</span><span class="s">size"</span><span class="pi">:</span> <span class="nv">176552</span><span class="pi">,</span>
            <span class="pi">}</span>
        <span class="pi">]</span>
    <span class="pi">},</span>
    <span class="s2">"</span><span class="s">qc_html"</span><span class="pi">:</span> <span class="pi">{</span>
        <span class="s2">"</span><span class="s">location"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">file:///home/username/rnaseq-cwl-training-exercises/Mov10_oe_1.subset_fastqc.html"</span><span class="pi">,</span>
        <span class="s2">"</span><span class="s">basename"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">Mov10_oe_1.subset_fastqc.html"</span><span class="pi">,</span>
        <span class="s2">"</span><span class="s">class"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">File"</span><span class="pi">,</span>
        <span class="s2">"</span><span class="s">size"</span><span class="pi">:</span> <span class="nv">383589</span>
    <span class="pi">}</span>
<span class="pi">}</span>
</code></pre></div></div>
</section>
  <section id="section-arvados-output">
<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="pi">{</span>
    <span class="s2">"</span><span class="s">bam_sorted_indexed"</span><span class="pi">:</span> <span class="pi">{</span>
        <span class="s2">"</span><span class="s">basename"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">Aligned.sortedByCoord.out.bam"</span><span class="pi">,</span>
        <span class="s2">"</span><span class="s">class"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">File"</span><span class="pi">,</span>
        <span class="s2">"</span><span class="s">location"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">keep:2dbaaef5aefd558e37f14280e47091a9+327/Aligned.sortedByCoord.out.bam"</span><span class="pi">,</span>
        <span class="s2">"</span><span class="s">secondaryFiles"</span><span class="pi">:</span> <span class="pi">[</span>
            <span class="pi">{</span>
                <span class="s2">"</span><span class="s">basename"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">Aligned.sortedByCoord.out.bam.bai"</span><span class="pi">,</span>
                <span class="s2">"</span><span class="s">class"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">File"</span><span class="pi">,</span>
                <span class="s2">"</span><span class="s">location"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">keep:2dbaaef5aefd558e37f14280e47091a9+327/Aligned.sortedByCoord.out.bam.bai"</span>
            <span class="pi">}</span>
        <span class="pi">],</span>
        <span class="s2">"</span><span class="s">size"</span><span class="pi">:</span> <span class="nv">25370695</span>
    <span class="pi">},</span>
    <span class="s2">"</span><span class="s">qc_html"</span><span class="pi">:</span> <span class="pi">{</span>
        <span class="s2">"</span><span class="s">basename"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">Mov10_oe_1.subset_fastqc.html"</span><span class="pi">,</span>
        <span class="s2">"</span><span class="s">class"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">File"</span><span class="pi">,</span>
        <span class="s2">"</span><span class="s">location"</span><span class="pi">:</span> <span class="s2">"</span><span class="s">keep:2dbaaef5aefd558e37f14280e47091a9+327/Mov10_oe_1.subset_fastqc.html"</span><span class="pi">,</span>
        <span class="s2">"</span><span class="s">size"</span><span class="pi">:</span> <span class="nv">383589</span>
    <span class="pi">}</span>
<span class="pi">}</span>
</code></pre></div></div>
</section>
</div>

<p>This has a similar structure as <code class="language-plaintext highlighter-rouge">main-input.yaml</code>.  The each output
parameter is listed, with the <code class="language-plaintext highlighter-rouge">location</code> field of each <code class="language-plaintext highlighter-rouge">File</code> object
indicating where the output file can be found.</p>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>The input parameter file is a YAML file with values for each input parameter.</p>
</li>
    
    <li><p>A common reason for a workflow step fails is insufficient RAM.</p>
</li>
    
    <li><p>Use ResourceRequirement to set the amount of RAM to be allocated to the job.</p>
</li>
    
    <li><p>Output parameter values are printed as JSON to standard output at the end of the run.</p>
</li>
    
  </ul>
</blockquote>

<hr />

<h1 id="writing-a-tool-wrapper" class="maintitle">Writing a Tool Wrapper</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 20 min
      <br />
      <strong>Exercises:</strong> 30 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>What are the key components of a tool wrapper?</p>
</li>
	
	<li><p>How do I use software containers to supply the software I want to run?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Write a tool wrapper for the featureCounts tool.</p>
</li>
	
	<li><p>Find an software container that has the software we want to use.</p>
</li>
	
	<li><p>Add the tool wrapper to our main workflow.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<p>This lesson has migrated to https://doc.arvados.org/rnaseq-cwl-training/04-commandlinetool/index.html</p>

<p>It is time to add the last step in the analysis.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Count mapped reads</span>
featureCounts <span class="nt">-T</span> <span class="nv">$cores</span> <span class="nt">-s</span> 2 <span class="nt">-a</span> <span class="nv">$gtf</span> <span class="nt">-o</span> <span class="nv">$counts</span> <span class="nv">$counts_input_bam</span>
</code></pre></div></div>

<p>This will use the “featureCounts” tool from the “subread” package.</p>

<h1 id="file-header">File header</h1>

<p>A CommandLineTool describes a single invocation of a command line
program.  It consumes some input parameters, runs a program, and
captures output, mainly in in the form of files produced by the
program.</p>

<p>Create a new file “featureCounts.cwl”</p>

<p>Let’s start with the header.  This is very similar to the workflow, except that we use <code class="language-plaintext highlighter-rouge">class: CommandLineTool</code>.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">cwlVersion</span><span class="pi">:</span> <span class="s">v1.2</span>
<span class="na">class</span><span class="pi">:</span> <span class="s">CommandLineTool</span>
<span class="na">label</span><span class="pi">:</span> <span class="s">featureCounts tool</span>
</code></pre></div></div>

<h1 id="command-line-tool-inputs">Command line tool inputs</h1>

<p>The <code class="language-plaintext highlighter-rouge">inputs</code> section describes input parameters with the same form as
the Workflow <code class="language-plaintext highlighter-rouge">inputs</code> section.</p>

<blockquote class="challenge">
  <h2 id="exercise">Exercise</h2>

  <p>The variables used in the bash script are <code class="language-plaintext highlighter-rouge">$cores</code>, <code class="language-plaintext highlighter-rouge">$gtf</code>, <code class="language-plaintext highlighter-rouge">$counts</code> and <code class="language-plaintext highlighter-rouge">$counts_input_bam</code>.</p>

  <ul>
    <li>$cores is the number of CPU cores to use.</li>
    <li>$gtf is the input .gtf file</li>
    <li>$counts is the name we will give to the output file</li>
    <li>$counts_input_bam is the input .bam file</li>
  </ul>

  <p>Write the <code class="language-plaintext highlighter-rouge">inputs</code> section for the File inputs <code class="language-plaintext highlighter-rouge">gtf</code> and <code class="language-plaintext highlighter-rouge">counts_input_bam</code>.</p>

  <blockquote class="solution">
    <h2 id="solution">Solution</h2>
    <div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">inputs</span><span class="pi">:</span>
  <span class="na">gtf</span><span class="pi">:</span> <span class="s">File</span>
  <span class="na">counts_input_bam</span><span class="pi">:</span> <span class="s">File</span>
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<h1 id="specifying-the-program-to-run">Specifying the program to run</h1>

<p>Give the name of the program to run in <code class="language-plaintext highlighter-rouge">baseCommand</code>.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">baseCommand</span><span class="pi">:</span> <span class="s">featureCounts</span>
</code></pre></div></div>

<h1 id="command-arguments">Command arguments</h1>

<p>The easiest way to describe the command line is with an <code class="language-plaintext highlighter-rouge">arguments</code>
section.  This takes a comma-separated list of command line arguments.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">arguments</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">-T</span><span class="pi">,</span> <span class="nv">$(runtime.cores)</span><span class="pi">,</span>
            <span class="nv">-a</span><span class="pi">,</span> <span class="nv">$(inputs.gtf)</span><span class="pi">,</span>
            <span class="nv">-o</span><span class="pi">,</span> <span class="nv">featurecounts.tsv</span><span class="pi">,</span>
            <span class="nv">$(inputs.counts_input_bam)</span><span class="pi">]</span>
</code></pre></div></div>

<p>Input variables are included on the command line as
<code class="language-plaintext highlighter-rouge">$(inputs.name_of_parameter)</code>.  When the tool is executed, the
variables will be replaced with the input parameter values.</p>

<p>There are also some special variables.  The <code class="language-plaintext highlighter-rouge">runtime</code> object describes
the resources allocated to running the program.  Here we use
<code class="language-plaintext highlighter-rouge">$(runtime.cores)</code> to decide how many threads to request.</p>

<blockquote class="callout">
  <h2 id="arguments-vs-inputbinding"><code class="language-plaintext highlighter-rouge">arguments</code> vs <code class="language-plaintext highlighter-rouge">inputBinding</code></h2>

  <p>You may recall from examining existing the fastqc and STAR tools
wrappers in lesson 2, another way to express command line parameters
is with <code class="language-plaintext highlighter-rouge">inputBinding</code> and <code class="language-plaintext highlighter-rouge">prefix</code> on individual input parameters.</p>

  <div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">inputs</span><span class="pi">:</span>
  <span class="na">parametername</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">parametertype</span>
    <span class="na">inputBinding</span><span class="pi">:</span>
      <span class="na">prefix</span><span class="pi">:</span> <span class="s">--some-option</span>
</code></pre></div>  </div>

  <p>We use <code class="language-plaintext highlighter-rouge">arguments</code> in the example simply because it is easier to see
how it lines up with the source shell script.</p>

  <p>You can use both <code class="language-plaintext highlighter-rouge">inputBinding</code> and <code class="language-plaintext highlighter-rouge">arguments</code> in the same
CommandLineTool document.  There is no “right” or “wrong” way, and
one does not override the other, they are combined to produce the
final command line invocation.</p>

</blockquote>

<h1 id="outputs-section">Outputs section</h1>

<p>In CWL, you must explicitly identify the outputs of a program.  This
associates output parameters with specific files, and enables the
workflow runner to know which files must be saved and which files can
be discarded.</p>

<p>In the previous section, we told the featureCounts program the name of
our output files should be <code class="language-plaintext highlighter-rouge">featurecounts.tsv</code>.</p>

<p>We can declare an output parameter called <code class="language-plaintext highlighter-rouge">featurecounts</code> that will
have that output file as its value.</p>

<p>The <code class="language-plaintext highlighter-rouge">outputBinding</code> section describes how to determine the value of
the parameter.  The <code class="language-plaintext highlighter-rouge">glob</code> field tells it to search for a file in the
output directory called <code class="language-plaintext highlighter-rouge">featurecounts.tsv</code></p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">outputs</span><span class="pi">:</span>
  <span class="na">featurecounts</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">outputBinding</span><span class="pi">:</span>
      <span class="na">glob</span><span class="pi">:</span> <span class="s">featurecounts.tsv</span>
</code></pre></div></div>

<h1 id="running-in-a-container">Running in a container</h1>

<p>In order to run the tool, it needs to be installed.
Using software containers, a tool can be pre-installed into a
compatible runtime environment, and that runtime environment (called a
container image) can be downloaded and run on demand.</p>

<p>Although plain CWL does not <em>require</em> the use of containers, many
popular platforms that run CWL do require the software be supplied in
the form of a container image.</p>

<blockquote class="challenge">
  <h2 id="finding-container-images">Finding container images</h2>

  <p>Many bioinformatics tools are already available as containers.  One
resource is the BioContainers project.  Let’s find the “subread” software:</p>

  <ol>
    <li>Visit <a href="https://biocontainers.pro/">https://biocontainers.pro/</a></li>
    <li>Click on “Registry”</li>
    <li>Search for “subread”</li>
    <li>Click on the search result for “subread”</li>
    <li>Click on the tab “Packages and Containers”</li>
    <li>Choose a row with type “docker”, then on the right side of the “Full
Tag” column for that row, click the “copy to clipboard” button.</li>
  </ol>

  <p>To declare that you want to run inside a container, add a section
called <code class="language-plaintext highlighter-rouge">hints</code> to your tool document.  Under <code class="language-plaintext highlighter-rouge">hints</code> add a
subsection <code class="language-plaintext highlighter-rouge">DockerRequirement</code>.  Under <code class="language-plaintext highlighter-rouge">DockerRequirement</code>, paste
the text your copied in the above step.  Replace the text <code class="language-plaintext highlighter-rouge">docker
pull</code> to <code class="language-plaintext highlighter-rouge">dockerPull:</code> ensure it is indented twice so it is a field
of <code class="language-plaintext highlighter-rouge">DockerRequirement</code>.</p>

  <blockquote class="solution">
    <h2 id="answer">Answer</h2>
    <div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">hints</span><span class="pi">:</span>
  <span class="na">DockerRequirement</span><span class="pi">:</span>
    <span class="na">dockerPull</span><span class="pi">:</span> <span class="s">quay.io/biocontainers/subread:1.5.0p3--0</span>
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<h1 id="running-a-tool-on-its-own">Running a tool on its own</h1>

<p>When creating a tool wrapper, it is helpful to run it on its own to test it.</p>

<p>The input to a single tool is the same kind of input parameters file
that we used as input to a workflow in the previous lesson.</p>

<p><code class="language-plaintext highlighter-rouge">featureCounts.yaml</code></p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">counts_input_bam</span><span class="pi">:</span>
  <span class="na">class</span><span class="pi">:</span> <span class="s">File</span>
  <span class="na">location</span><span class="pi">:</span> <span class="s">Aligned.sortedByCoord.out.bam</span>
<span class="na">gtf</span><span class="pi">:</span>
  <span class="na">class</span><span class="pi">:</span> <span class="s">File</span>
  <span class="na">location</span><span class="pi">:</span> <span class="s">rnaseq/reference_data/chr1-hg19_genes.gtf</span>
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="running-the-tool">Running the tool</h2>

  <p>Run the tool on its own to confirm it has correct behavior:</p>

  <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cwl-runner featureCounts.cwl featureCounts.yaml
</code></pre></div>  </div>
</blockquote>

<h1 id="adding-it-to-the-workflow">Adding it to the workflow</h1>

<p>Now that we have confirmed that the tool wrapper works, it is time to
add it to our workflow.</p>

<blockquote class="challenge">
  <h2 id="exercise-1">Exercise</h2>

  <ol>
    <li>Add a new step called <code class="language-plaintext highlighter-rouge">featureCounts</code> that runs our tool
wrapper.  The new step should take input from
<code class="language-plaintext highlighter-rouge">samtools/bam_sorted_indexed</code>, and should be allocated a
minimum of 500 MB of RAM</li>
    <li>Add a new output parameter for the workflow called
<code class="language-plaintext highlighter-rouge">featurecounts</code> The output source should come from the output
of the new <code class="language-plaintext highlighter-rouge">featureCounts</code> step.</li>
    <li>When you have an answer, run the updated workflow, which
should run the “featureCounts” step and produce “featurecounts”
output parameter.</li>
  </ol>

  <blockquote class="solution">
    <h2 id="answer-1">Answer</h2>
    <div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">steps</span><span class="pi">:</span>
  <span class="s">...</span>
  <span class="s">featureCounts</span><span class="pi">:</span>
    <span class="na">requirements</span><span class="pi">:</span>
      <span class="na">ResourceRequirement</span><span class="pi">:</span>
        <span class="na">ramMin</span><span class="pi">:</span> <span class="m">500</span>
    <span class="na">run</span><span class="pi">:</span> <span class="s">featureCounts.cwl</span>
    <span class="na">in</span><span class="pi">:</span>
      <span class="na">counts_input_bam</span><span class="pi">:</span> <span class="s">samtools/bam_sorted_indexed</span>
      <span class="na">gtf</span><span class="pi">:</span> <span class="s">gtf</span>
    <span class="na">out</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">featurecounts</span><span class="pi">]</span>

<span class="na">outputs</span><span class="pi">:</span>
  <span class="s">...</span>
  <span class="s">featurecounts</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">outputSource</span><span class="pi">:</span> <span class="s">featureCounts/featurecounts</span>
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<blockquote class="solution">
  <h2 id="episode-solution">Episode solution</h2>
  <ul>
    <li><a href="../assets/answers/ep4/main.cwl">main.cwl</a></li>
    <li><a href="../assets/answers/ep4/featureCounts.cwl">featureCounts.cwl</a></li>
  </ul>
</blockquote>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>The key components of a command line tool wrapper are the header, inputs, baseCommand, arguments, and outputs.</p>
</li>
    
    <li><p>Like workflows, CommandLineTools have <code class="language-plaintext highlighter-rouge">inputs</code> and <code class="language-plaintext highlighter-rouge">outputs</code>.</p>
</li>
    
    <li><p>Use <code class="language-plaintext highlighter-rouge">baseCommand</code> and <code class="language-plaintext highlighter-rouge">arguments</code> to provide the program to run and the command line arguments to run it with.</p>
</li>
    
    <li><p>Use <code class="language-plaintext highlighter-rouge">glob</code> to capture output files and assign them to output parameters.</p>
</li>
    
    <li><p>Use DockerRequirement to supply the name of the Docker image that contains the software to run.</p>
</li>
    
  </ul>
</blockquote>

<hr />

<h1 id="analyzing-multiple-samples" class="maintitle">Analyzing Multiple Samples</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 30 min
      <br />
      <strong>Exercises:</strong> 30 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>How can you run the same workflow over multiple samples?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Modify the workflow to process multiple samples, then perform a joint analysis.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<p>This lesson has migrated to https://doc.arvados.org/rnaseq-cwl-training/05-scatter/index.html</p>

<p>In the previous lesson, we completed converting the function of the
original source shell script into CWL.  This lesson expands the scope
by demonstrating what changes to make to the workflow to be able to
analyze multiple samples in parallel.</p>

<h1 id="subworkflows">Subworkflows</h1>

<p>In addition to running command line tools, a workflow step can also
execute another workflow.</p>

<p>First, copy <code class="language-plaintext highlighter-rouge">main.cwl</code> to <code class="language-plaintext highlighter-rouge">alignment.cwl</code>.</p>

<p>Next, open <code class="language-plaintext highlighter-rouge">main.cwl</code> for editing.  We are going to replace the <code class="language-plaintext highlighter-rouge">steps</code> and <code class="language-plaintext highlighter-rouge">outputs</code> sections.</p>

<p>Remove all the steps and replace them with a single <code class="language-plaintext highlighter-rouge">alignment</code> step
which invokes the <code class="language-plaintext highlighter-rouge">alignment.cwl</code> we just copied.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">steps</span><span class="pi">:</span>
  <span class="na">alignment</span><span class="pi">:</span>
    <span class="na">run</span><span class="pi">:</span> <span class="s">alignment.cwl</span>
    <span class="na">in</span><span class="pi">:</span>
      <span class="na">fq</span><span class="pi">:</span> <span class="s">fq</span>
      <span class="na">genome</span><span class="pi">:</span> <span class="s">genome</span>
      <span class="na">gtf</span><span class="pi">:</span> <span class="s">gtf</span>
    <span class="na">out</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">qc_html</span><span class="pi">,</span> <span class="nv">bam_sorted_indexed</span><span class="pi">,</span> <span class="nv">featurecounts</span><span class="pi">]</span>
</code></pre></div></div>

<p>In the <code class="language-plaintext highlighter-rouge">outputs</code> section, all the output sources are from the alignment step:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">outputs</span><span class="pi">:</span>
  <span class="na">qc_html</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">outputSource</span><span class="pi">:</span> <span class="s">alignment/qc_html</span>
  <span class="na">bam_sorted_indexed</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">outputSource</span><span class="pi">:</span> <span class="s">alignment/bam_sorted_indexed</span>
  <span class="na">featurecounts</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">outputSource</span><span class="pi">:</span> <span class="s">alignment/featurecounts</span>
</code></pre></div></div>

<p>We also need add “SubworkflowFeatureRequirement” to tell the workflow
runner that we are using subworkflows:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">requirements</span><span class="pi">:</span>
  <span class="na">SubworkflowFeatureRequirement</span><span class="pi">:</span> <span class="pi">{}</span>
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="running-the-workflow">Running the workflow</h2>

  <p>Run this workflow.  You should get exactly the same results as
before, as all we have done so far is to wrap the inner workflow with
an outer workflow.</p>

</blockquote>

<blockquote class="solution">
  <h2 id="part-1-solution">Part 1 solution</h2>
  <ul>
    <li><a href="../assets/answers/ep5/part1/main.cwl">main.cwl</a></li>
    <li><a href="../assets/answers/ep5/part1/alignment.cwl">alignment.cwl</a></li>
    <li><a href="../assets/answers/ep5/part1/featureCounts.cwl">featureCounts.cwl</a></li>
  </ul>
</blockquote>

<h1 id="scattering">Scattering</h1>

<p>The “wrapper” step lets us do something useful.  We can modify the
outer workflow to accept a list of files, and then invoke the inner
workflow step for every one of those files.  We will need to modify
the <code class="language-plaintext highlighter-rouge">inputs</code>, <code class="language-plaintext highlighter-rouge">steps</code>, <code class="language-plaintext highlighter-rouge">outputs</code>, and <code class="language-plaintext highlighter-rouge">requirements</code> sections.</p>

<p>First we change the <code class="language-plaintext highlighter-rouge">fq</code> parameter to expect a list of files:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">inputs</span><span class="pi">:</span>
  <span class="na">fq</span><span class="pi">:</span> <span class="s">File[]</span>
  <span class="na">genome</span><span class="pi">:</span> <span class="s">Directory</span>
  <span class="na">gtf</span><span class="pi">:</span> <span class="s">File</span>
</code></pre></div></div>

<p>Next, we add <code class="language-plaintext highlighter-rouge">scatter</code> to the alignment step.  The means we want to
run run <code class="language-plaintext highlighter-rouge">alignment.cwl</code> for each value in the list in the <code class="language-plaintext highlighter-rouge">fq</code>
parameter.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">steps</span><span class="pi">:</span>
  <span class="na">alignment</span><span class="pi">:</span>
    <span class="na">run</span><span class="pi">:</span> <span class="s">alignment.cwl</span>
    <span class="na">scatter</span><span class="pi">:</span> <span class="s">fq</span>
    <span class="na">in</span><span class="pi">:</span>
      <span class="na">fq</span><span class="pi">:</span> <span class="s">fq</span>
      <span class="na">genome</span><span class="pi">:</span> <span class="s">genome</span>
      <span class="na">gtf</span><span class="pi">:</span> <span class="s">gtf</span>
    <span class="na">out</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">qc_html</span><span class="pi">,</span> <span class="nv">bam_sorted_indexed</span><span class="pi">,</span> <span class="nv">featurecounts</span><span class="pi">]</span>
</code></pre></div></div>

<p>Because the scatter produces multiple outputs, each output parameter
becomes a list as well:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">outputs</span><span class="pi">:</span>
  <span class="na">qc_html</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">File[]</span>
    <span class="na">outputSource</span><span class="pi">:</span> <span class="s">alignment/qc_html</span>
  <span class="na">bam_sorted_indexed</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">File[]</span>
    <span class="na">outputSource</span><span class="pi">:</span> <span class="s">alignment/bam_sorted_indexed</span>
  <span class="na">featurecounts</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">File[]</span>
    <span class="na">outputSource</span><span class="pi">:</span> <span class="s">alignment/featurecounts</span>
</code></pre></div></div>

<p>We also need add “ScatterFeatureRequirement” to tell the workflow
runner that we are using scatter:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">requirements</span><span class="pi">:</span>
  <span class="na">SubworkflowFeatureRequirement</span><span class="pi">:</span> <span class="pi">{}</span>
  <span class="na">ScatterFeatureRequirement</span><span class="pi">:</span> <span class="pi">{}</span>
</code></pre></div></div>

<blockquote class="solution">
  <h2 id="part-2-solution">Part 2 solution</h2>
  <ul>
    <li><a href="../assets/answers/ep5/part2/main.cwl">main.cwl</a></li>
    <li><a href="../assets/answers/ep5/part2/alignment.cwl">alignment.cwl</a></li>
    <li><a href="../assets/answers/ep5/part2/featureCounts.cwl">featureCounts.cwl</a></li>
  </ul>
</blockquote>

<h1 id="input-parameter-lists">Input parameter lists</h1>

<p>The <code class="language-plaintext highlighter-rouge">fq</code> parameter needs to be a list.  You write a list in yaml by
starting each list item with a dash.  Example <code class="language-plaintext highlighter-rouge">main-input.yaml</code></p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">fq</span><span class="pi">:</span>
  <span class="pi">-</span> <span class="na">class</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">location</span><span class="pi">:</span> <span class="s">rnaseq/raw_fastq/Mov10_oe_1.subset.fq</span>
    <span class="na">format</span><span class="pi">:</span> <span class="s">http://edamontology.org/format_1930</span>
  <span class="pi">-</span> <span class="na">class</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">location</span><span class="pi">:</span> <span class="s">rnaseq/raw_fastq/Mov10_oe_2.subset.fq</span>
    <span class="na">format</span><span class="pi">:</span> <span class="s">http://edamontology.org/format_1930</span>
  <span class="pi">-</span> <span class="na">class</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">location</span><span class="pi">:</span> <span class="s">rnaseq/raw_fastq/Mov10_oe_3.subset.fq</span>
    <span class="na">format</span><span class="pi">:</span> <span class="s">http://edamontology.org/format_1930</span>
  <span class="pi">-</span> <span class="na">class</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">location</span><span class="pi">:</span> <span class="s">rnaseq/raw_fastq/Irrel_kd_1.subset.fq</span>
    <span class="na">format</span><span class="pi">:</span> <span class="s">http://edamontology.org/format_1930</span>
  <span class="pi">-</span> <span class="na">class</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">location</span><span class="pi">:</span> <span class="s">rnaseq/raw_fastq/Irrel_kd_2.subset.fq</span>
    <span class="na">format</span><span class="pi">:</span> <span class="s">http://edamontology.org/format_1930</span>
  <span class="pi">-</span> <span class="na">class</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">location</span><span class="pi">:</span> <span class="s">rnaseq/raw_fastq/Irrel_kd_3.subset.fq</span>
    <span class="na">format</span><span class="pi">:</span> <span class="s">http://edamontology.org/format_1930</span>
<span class="na">genome</span><span class="pi">:</span>
  <span class="na">class</span><span class="pi">:</span> <span class="s">Directory</span>
  <span class="na">location</span><span class="pi">:</span> <span class="s">hg19-chr1-STAR-index</span>
<span class="na">gtf</span><span class="pi">:</span>
  <span class="na">class</span><span class="pi">:</span> <span class="s">File</span>
  <span class="na">location</span><span class="pi">:</span> <span class="s">rnaseq/reference_data/chr1-hg19_genes.gtf</span>
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="running-the-workflow-1">Running the workflow</h2>

  <p>Run this workflow.  You will now get results for each one of the
input fastq files.</p>

</blockquote>

<h1 id="combining-results">Combining results</h1>

<p>Each instance of the alignment workflow produces its own
<code class="language-plaintext highlighter-rouge">featurecounts.tsv</code> file.  However, to be able to compare results
easily, we would like single file with all the results.</p>

<p>We can modify the workflow to run <code class="language-plaintext highlighter-rouge">featureCounts</code> once at the end of
the workflow, taking all the bam files listed on the command line.</p>

<p>We will need to change a few things.</p>

<p>First, in <code class="language-plaintext highlighter-rouge">featureCounts.cwl</code> we need to modify it to accept either a
single bam file or list of bam files.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">inputs</span><span class="pi">:</span>
  <span class="na">gtf</span><span class="pi">:</span> <span class="s">File</span>
  <span class="na">counts_input_bam</span><span class="pi">:</span>
   <span class="pi">-</span> <span class="s">File</span>
   <span class="pi">-</span> <span class="s">File[]</span>
</code></pre></div></div>

<p>Second, in <code class="language-plaintext highlighter-rouge">alignment.cwl</code> we need to remove the <code class="language-plaintext highlighter-rouge">featureCounts</code> step from alignment.cwl, as well as the <code class="language-plaintext highlighter-rouge">featurecounts</code> output parameter.</p>

<p>Third, in <code class="language-plaintext highlighter-rouge">main.cwl</code> we need to remove <code class="language-plaintext highlighter-rouge">featurecounts</code> from the <code class="language-plaintext highlighter-rouge">alignment</code> step
outputs, and add a new step:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">steps</span><span class="pi">:</span>
  <span class="na">alignment</span><span class="pi">:</span>
    <span class="na">run</span><span class="pi">:</span> <span class="s">alignment.cwl</span>
    <span class="na">scatter</span><span class="pi">:</span> <span class="s">fq</span>
    <span class="na">in</span><span class="pi">:</span>
      <span class="na">fq</span><span class="pi">:</span> <span class="s">fq</span>
      <span class="na">genome</span><span class="pi">:</span> <span class="s">genome</span>
      <span class="na">gtf</span><span class="pi">:</span> <span class="s">gtf</span>
    <span class="na">out</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">qc_html</span><span class="pi">,</span> <span class="nv">bam_sorted_indexed</span><span class="pi">]</span>
  <span class="na">featureCounts</span><span class="pi">:</span>
    <span class="na">requirements</span><span class="pi">:</span>
      <span class="na">ResourceRequirement</span><span class="pi">:</span>
        <span class="na">ramMin</span><span class="pi">:</span> <span class="m">500</span>
    <span class="na">run</span><span class="pi">:</span> <span class="s">featureCounts.cwl</span>
    <span class="na">in</span><span class="pi">:</span>
      <span class="na">counts_input_bam</span><span class="pi">:</span> <span class="s">alignment/bam_sorted_indexed</span>
      <span class="na">gtf</span><span class="pi">:</span> <span class="s">gtf</span>
    <span class="na">out</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">featurecounts</span><span class="pi">]</span>
</code></pre></div></div>

<p>Last, we modify the <code class="language-plaintext highlighter-rouge">featurecounts</code> output parameter.  Instead of a
list of files produced by the <code class="language-plaintext highlighter-rouge">alignment</code> step, it is now a single
file produced by the new <code class="language-plaintext highlighter-rouge">featureCounts</code> step.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">outputs</span><span class="pi">:</span>
  <span class="s">...</span>
  <span class="s">featurecounts</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">outputSource</span><span class="pi">:</span> <span class="s">featureCounts/featurecounts</span>
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="running-the-workflow-2">Running the workflow</h2>

  <p>Run this workflow.  You will still have separate results from fastq
and and STAR, but now you will only have a single
<code class="language-plaintext highlighter-rouge">featurecounts.tsv</code> file with a column for each bam file.</p>

</blockquote>

<blockquote class="solution">
  <h2 id="episode-solution">Episode solution</h2>
  <ul>
    <li><a href="../assets/answers/ep5/part4/main.cwl">main.cwl</a></li>
    <li><a href="../assets/answers/ep5/part4/alignment.cwl">alignment.cwl</a></li>
    <li><a href="../assets/answers/ep5/part4/featureCounts.cwl">featureCounts.cwl</a></li>
  </ul>
</blockquote>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>Separate the part of the workflow that you want to run multiple times into a subworkflow.</p>
</li>
    
    <li><p>Use a scatter step to run the subworkflow over a list of inputs.</p>
</li>
    
    <li><p>The result of a scatter is an array, which can be used in a combine step to get a single result.</p>
</li>
    
  </ul>
</blockquote>

<hr />

<h1 id="dynamic-workflow-behavior" class="maintitle">Dynamic Workflow Behavior</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 20 min
      <br />
      <strong>Exercises:</strong> 10 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>What kind of custom logic can happen between steps?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Customize the STAR output filename to use the input filename.</p>
</li>
	
	<li><p>Organize files into directories.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<p>This lesson has migrated to https://doc.arvados.org/rnaseq-cwl-training/06-expressions/index.html</p>

<h1 id="expressions-on-step-inputs">Expressions on step inputs</h1>

<p>You might have noticed that the output bam files are all named
<code class="language-plaintext highlighter-rouge">Aligned.sortedByCoord.out.bam</code>.  This happens because because when we
call STAR, it gives the output a default file name.</p>

<p>During workflow execution, this is usually not a problem.  The
workflow runner is smart enough to know that these files are different
and keep them separate.  This can even make development easier by not
having to worry about assigning unique file names to every file.
Also, if we intend to discard the BAM files as intermediate results</p>

<p>However, it is a problem for humans interpreting the output.  We can
fix this by setting the parameter <code class="language-plaintext highlighter-rouge">OutFileNamePrefix</code> on STAR.  We
want the output filename to be based on the input filename.</p>

<p>In <code class="language-plaintext highlighter-rouge">alignment.cwl</code>, we can use <code class="language-plaintext highlighter-rouge">valueFrom</code> on the <code class="language-plaintext highlighter-rouge">OutFileNamePrefix</code>
input parameter to construct the output prefix from the input
filename.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">requirements</span><span class="pi">:</span>
  <span class="na">StepInputExpressionRequirement</span><span class="pi">:</span> <span class="pi">{}</span>

<span class="na">steps</span><span class="pi">:</span>
  <span class="s">...</span>
  <span class="s">STAR</span><span class="pi">:</span>
    <span class="s">...</span>
    <span class="s">in</span><span class="pi">:</span>
      <span class="na">ForwardReads</span><span class="pi">:</span> <span class="s">fq</span>
      <span class="s">...</span>
      <span class="na">OutFileNamePrefix</span><span class="pi">:</span> <span class="pi">{</span><span class="nv">valueFrom</span><span class="pi">:</span> <span class="s2">"</span><span class="s">$(inputs.ForwardReads.nameroot)."</span><span class="pi">}</span>
</code></pre></div></div>

<p>The code between <code class="language-plaintext highlighter-rouge">$(...)</code> is called an “expression”.  It is evaluated
when setting up the step to run, and the expression is replaced by the
result to get the parameter value.</p>

<p>An expression can refer to other inputs to the step that are either
directly connected to another value, or have a default value.  Here,
we refer to the input parameter ForwardReads, which is our fastq input
file.</p>

<p>ForwardReads is a File object, not a plain file name, so it has some
fields of its own.  The file object has a number of fields that we can
use.  These include <code class="language-plaintext highlighter-rouge">basename</code> (the name of the file, without a
directory), <code class="language-plaintext highlighter-rouge">size</code> (file size, in bytes), <code class="language-plaintext highlighter-rouge">nameext</code> (the last file
extension) and <code class="language-plaintext highlighter-rouge">nameroot</code> (the name with <code class="language-plaintext highlighter-rouge">nameext</code> removed).  Using</p>

<p>Finally, our expression is embedded in a string, so after replacing
the expression with the value of <code class="language-plaintext highlighter-rouge">inputs.ForwardReads.nameroot</code>, it
adds the remainder of the string, which just is a dot <code class="language-plaintext highlighter-rouge">.</code>.  This is to
separate the leading part of our filename from the “Aligned.bam”
extension that will be added by STAR.</p>

<h1 id="organizing-output-files-into-directories">Organizing output files into Directories</h1>

<p>You probably noticed that all the output files appear in the same
directory.  You might prefer that each file appears in its own
directory.  This will show you how to do that.</p>

<p>Unlike shell scripts, in CWL you cannot call <code class="language-plaintext highlighter-rouge">mkdir</code> and <code class="language-plaintext highlighter-rouge">mv</code> to
organize files into directories.  This is because the output files, at
this point, do not actually exist together in one directory.  They may
exist on different nodes, in different directories, or different cloud
buckets.  In CWL, instead of moving files around directly, you tell
the runner you want your directory to look like, and it will create it
for you.</p>

<p>We can use an “expression” to create a <code class="language-plaintext highlighter-rouge">Directory</code> object describing
each of our directories.  An expression is a piece of Javascript code
which is executed by the workflow runner to produce values that affect
the workflow.  These can be a simple as substituting the value of an
input variable, or as complex as en entire function that generates new
objects.</p>

<p>Javscript code must be bracketed inside <code class="language-plaintext highlighter-rouge">$(...)</code> or <code class="language-plaintext highlighter-rouge">${...}</code>. The
difference comes down to syntax.  The <code class="language-plaintext highlighter-rouge">$()</code> form is more compact but
can only include code that can appear on the right side of an
assignment (<code class="language-plaintext highlighter-rouge">=</code>), which cannot include control blocks like <code class="language-plaintext highlighter-rouge">if</code> or
<code class="language-plaintext highlighter-rouge">for</code>.  The <code class="language-plaintext highlighter-rouge">${}</code> form is a Javascript function, which can include
control blocks, and must end in a <code class="language-plaintext highlighter-rouge">return</code> statement.</p>

<p>Expressions can both appear in <code class="language-plaintext highlighter-rouge">valueFrom</code> fields as well as some
other fields, or in an <code class="language-plaintext highlighter-rouge">ExpressionTool</code> which, like <code class="language-plaintext highlighter-rouge">Workflow</code> or
<code class="language-plaintext highlighter-rouge">CommandLineTool</code> has explicitly defined <code class="language-plaintext highlighter-rouge">inputs</code> and <code class="language-plaintext highlighter-rouge">outputs</code>
sections.</p>

<p>The approach here is to define an expression tool which takes a</p>

<p>The <code class="language-plaintext highlighter-rouge">Directory</code> object has two fields, <code class="language-plaintext highlighter-rouge">basename</code> and <code class="language-plaintext highlighter-rouge">listing</code>.  The
<code class="language-plaintext highlighter-rouge">basename</code> is the name of the directory, and the <code class="language-plaintext highlighter-rouge">listing</code> is the
contents, which consists of other File and Directory objects.</p>

<p>Create <code class="language-plaintext highlighter-rouge">subdirs.cwl</code>:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">cwlVersion</span><span class="pi">:</span> <span class="s">v1.2</span>
<span class="na">class</span><span class="pi">:</span> <span class="s">ExpressionTool</span>
<span class="na">requirements</span><span class="pi">:</span>
  <span class="na">InlineJavascriptRequirement</span><span class="pi">:</span> <span class="pi">{}</span>
<span class="na">inputs</span><span class="pi">:</span>
  <span class="na">fq</span><span class="pi">:</span> <span class="s">File[]</span>
  <span class="na">bams</span><span class="pi">:</span> <span class="s">File[]</span>
  <span class="na">qc</span><span class="pi">:</span> <span class="s">File[]</span>
<span class="na">outputs</span><span class="pi">:</span>
  <span class="na">dirs</span><span class="pi">:</span> <span class="s">Directory[]</span>
<span class="na">expression</span><span class="pi">:</span> <span class="pi">|-</span>
  <span class="s">${</span>
  <span class="s">var dirs = [];</span>
  <span class="s">for (var i = 0; i &lt; inputs.bams.length; i++) {</span>
    <span class="s">dirs.push({</span>
      <span class="s">"class": "Directory",</span>
      <span class="s">"basename": inputs.fq[i].nameroot,</span>
      <span class="s">"listing": [inputs.bams[i], inputs.qc[i]]</span>
    <span class="s">});</span>
  <span class="s">}</span>
  <span class="s">return {"dirs": dirs};</span>
  <span class="s">}</span>
</code></pre></div></div>

<p>Then change <code class="language-plaintext highlighter-rouge">main.cwl</code>:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">steps</span><span class="pi">:</span>
  <span class="s">...</span>
  <span class="s">output-subdirs</span><span class="pi">:</span>
    <span class="na">run</span><span class="pi">:</span> <span class="s">subdirs.cwl</span>
    <span class="na">in</span><span class="pi">:</span>
      <span class="na">fq</span><span class="pi">:</span> <span class="s">fq</span>
      <span class="na">bams</span><span class="pi">:</span> <span class="s">alignment/bam_sorted_indexed</span>
      <span class="na">qc</span><span class="pi">:</span> <span class="s">alignment/qc_html</span>
    <span class="na">out</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">dirs</span><span class="pi">]</span>
<span class="na">outputs</span><span class="pi">:</span>
  <span class="na">dirs</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">Directory[]</span>
    <span class="na">outputSource</span><span class="pi">:</span> <span class="s">output-subdirs/dirs</span>
  <span class="na">featurecounts</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">outputSource</span><span class="pi">:</span> <span class="s">featureCounts/featurecounts</span>
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="running-the-workflow">Running the workflow</h2>

  <p>Run the workflow.  Look at the output.  The BAM and fastqc files
should now be organized into directories, with better naming of the
bam files.</p>

</blockquote>

<blockquote class="solution">
  <h2 id="episode-solution">Episode solution</h2>
  <ul>
    <li><a href="../assets/answers/ep6/main.cwl">main.cwl</a></li>
    <li><a href="../assets/answers/ep6/alignment.cwl">alignment.cwl</a></li>
    <li><a href="../assets/answers/ep6/featureCounts.cwl">featureCounts.cwl</a></li>
    <li><a href="../assets/answers/ep6/subdirs.cwl">subdirs.cwl</a></li>
  </ul>
</blockquote>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>CWL expressions allow you to use custom logic to determine input parameter values.</p>
</li>
    
    <li><p>CWL ExpressionTool can be used to reshape data, such as declaring directories that should contain output files.</p>
</li>
    
  </ul>
</blockquote>

<hr />

<h1 id="resources-for-further-learning" class="maintitle"> Resources for further learning</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 10 min
      <br />
      <strong>Exercises:</strong> 0 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>Where should I go to learn more?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Become a part of the CWL community.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<p>This lesson has migrated to https://doc.arvados.org/rnaseq-cwl-training/07-resources/index.html</p>

<p>Hopefully you now have a basic grasp of the steps involved in
developing a CWL workflow. There are many resources out there to
further help you use CWL to solve your own scientific workflow
problems.</p>

<h1 id="cwl-reference">CWL Reference</h1>

<p><a href="https://commonwl.org">Main CWL web page</a></p>

<p><a href="https://www.commonwl.org/user_guide/">User guide</a></p>

<p><a href="https://www.commonwl.org/v1.2/">Specification</a></p>

<p><a href="https://github.com/common-workflow-language/">Github organization</a></p>

<h1 id="cwl-community">CWL Community</h1>

<p>The <a href="https://cwl.discourse.group/">CWL Forum</a> is is best place to ask questions</p>

<p><a href="https://gitter.im/common-workflow-language/common-workflow-language">Gitter (chat)</a></p>

<p><a href="https://cwl.discourse.group/t/eu-us-timezone-cwl-video-chat/260">Weekly video calls</a></p>

<h1 id="software-resources">Software resources</h1>

<p>Github organization for <a href="https://github.com/common-workflow-library/">repositories of CWL tool and workflow descriptions</a>,
including <a href="https://github.com/common-workflow-library/bio-cwl-tools">bio-cwl-tools</a>.</p>

<p><a href="https://biocontainers.pro/">BioContainers</a></p>

<p><a href="https://github.com/search?q=extension%3Acwl+cwlVersion">Search for CWL files</a> on
Github, try adding the name of a tool you are interested in to the
search</p>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>Learn more advanced techniques from CWL user guide, by asking questions on the CWL forum and chat channel, and reading the specification.</p>
</li>
    
  </ul>
</blockquote>

<hr />

<h1 id="supplement-creating-docker-images-for-workflows" class="maintitle">Supplement: Creating Docker Images for Workflows</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 10 min
      <br />
      <strong>Exercises:</strong> 1 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>How do I create Docker images from scratch?</p>
</li>
	
	<li><p>What some best practices for Docker images?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Understand how to get started writing Dockerfiles</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<p>This lesson has migrated to https://doc.arvados.org/rnaseq-cwl-training/08-supplement-docker/index.html</p>

<p>Common Workflow Language supports running tasks inside software
containers.  Software container systems (such as Docker) create an
execution environment that is isolated from the host system, so that
software installed on the host system does not conflict with the
software installed inside the container.</p>

<p>Programs running inside a software container get a different (and
generally restricted) view of the system than processes running
outside the container.  One of the most important and useful features
is that the containerized program has a different view of the file
system.  A program running inside a container, searching for
libraries, modules, configuration files, data files, etc, only sees
the files defined inside the container.</p>

<p>This means that, usually, a given file <em>path</em> refers to <em>different
actual files</em> depending from the persective of being inside or outside
the container.  It is also possible to have a file from the host
system appear at some location inside the container, meaning that the
<em>same file</em> appears at <em>different paths</em> depending from the persective
of being inside or outside the container.</p>

<p>The complexity of translating between the container and its host
environment is handled by the Common Workflow Language runner.  As a
workflow author, you only need to worry about the environment <em>inside</em>
the container.</p>

<h1 id="what-are-docker-images">What are Docker images?</h1>

<p>The Docker image describes the starting conditions for the container.
Most importantly, this includes starting layout and contents of the
container’s file system.  This file system is typically a lightweight
POSIX environment, providing a standard set of POSIX utilities like a
<code class="language-plaintext highlighter-rouge">sh</code>, <code class="language-plaintext highlighter-rouge">ls</code>, <code class="language-plaintext highlighter-rouge">cat</code>, etc and organized into standard POSIX directories
like <code class="language-plaintext highlighter-rouge">/bin</code> and <code class="language-plaintext highlighter-rouge">/lib</code>.</p>

<p>The image is is made up of multiple “layers”.  Each layer modifies the
layer below it by adding, removing or modifying files to produce a new
layer.  This allows lower layers to be re-used.</p>

<h1 id="writing-a-dockerfile">Writing a Dockerfile</h1>

<p>In this example, we will build a Docker image containing the
Burrows-Wheeler Aligner (BWA) by Heng Li.  This is just for
demonstration, in practice you should prefer to use existing
containers from <a href="https://biocontainers.pro/">BioContainers</a>, which
includes <code class="language-plaintext highlighter-rouge">bwa</code>.</p>

<p>Each line of the Docker file consists of a COMMAND in all caps,
following by the parameters of that command.</p>

<p>The first line of the file will specify the base image that we are
going to build from.  As mentioned, images are divided up into
“layers”, so this tells Docker what to use for the first layer.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>FROM debian:10-slim
</code></pre></div></div>

<p>This starts from the lightweight (“slim”) Debian 10 Docker image.</p>

<p>Docker images have a special naming scheme.</p>

<p>A bare name like “debian” or “ubuntu” means it is an official Docker
image.  It has an implied prefix of “library”, so you may see the
image referred to as “library/debian”.  Official images are published
on <a href="https://hub.docker.com/search?type=image&amp;image_filter=official">Docker Hub</a>.</p>

<p>A name with two parts separated by a slash is published on Docker Hub
by someone else.  For example, <code class="language-plaintext highlighter-rouge">amazon/aws-cli</code> is published by
Amazon.  These can also be found on <a href="https://hub.docker.com/search?type=image">Docker Hub</a>.</p>

<p>A name with three parts separated by slashes means it is published on
a different container register.  For example,
<code class="language-plaintext highlighter-rouge">quay.io/biocontainers/subread</code> is published by <code class="language-plaintext highlighter-rouge">quay.io</code>.</p>

<p>Following image name, separated by a colon is the “tag”.  This is
typically the version of the image.  If not provided, the default tag
is “latest”.  In this example, the tag is “10-slim” indicating Debian
release 10.</p>

<p>The Docker file should also include a MAINTAINER (this is purely
metadata, it is stored in the image but not used for execution).</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>MAINTAINER Peter Amstutz &lt;peter.amstutz@curii.com&gt;
</code></pre></div></div>

<p>Next is the default user inside the image.  By making choosing root,
we can change anything inside the image (but not outside).</p>

<p>The body of the Dockerfile is a series of <code class="language-plaintext highlighter-rouge">RUN</code> commands.</p>

<p>Each command is run with <code class="language-plaintext highlighter-rouge">/bin/sh</code> inside the Docker container.</p>

<p>Each <code class="language-plaintext highlighter-rouge">RUN</code> command creates a new layer.</p>

<p>The <code class="language-plaintext highlighter-rouge">RUN</code> command can span multiple lines by using a trailing
backslash.</p>

<p>For the first command, we use <code class="language-plaintext highlighter-rouge">apt-get</code> to install some packages that
will be needed to compile <code class="language-plaintext highlighter-rouge">bwa</code>.  The <code class="language-plaintext highlighter-rouge">build-essential</code> package
installs <code class="language-plaintext highlighter-rouge">gcc</code>, <code class="language-plaintext highlighter-rouge">make</code>, etc.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>RUN apt-get update -qy &amp;&amp; \
	apt-get install -qy build-essential wget unzip
</code></pre></div></div>

<p>Now we do everything else: download the source code of bwa, unzip it,
make it, copy the resulting binary to <code class="language-plaintext highlighter-rouge">/usr/bin</code>, and clean up.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Install BWA 07.7.17
RUN wget https://github.com/lh3/bwa/archive/v0.7.17.zip &amp;&amp; \
	unzip v0.7.17 &amp;&amp; \
	cd bwa-0.7.17 &amp;&amp; \
	make &amp;&amp; \
	cp bwa /usr/bin &amp;&amp; \
	cd .. &amp;&amp; \
	rm -rf bwa-0.7.17
</code></pre></div></div>

<p>Because each <code class="language-plaintext highlighter-rouge">RUN</code> command creates a new layer, having the build and
clean up in separate <code class="language-plaintext highlighter-rouge">RUN</code> commands would mean creating a layer that
includes the intermediate object files from the build.  These would
then be carried around as part of the container image forever, despite
being useless.  By doing the entire build and clean up in one <code class="language-plaintext highlighter-rouge">RUN</code>
command, only the final state of the file system, with the binary
copied to <code class="language-plaintext highlighter-rouge">/usr/bin</code>, is committed to a layer.</p>

<p>To build a Docker image from a Dockerfile, use <code class="language-plaintext highlighter-rouge">docker build</code>.</p>

<p>Use the <code class="language-plaintext highlighter-rouge">-t</code> option to specify the name of the image.  Use <code class="language-plaintext highlighter-rouge">-f</code> if the
file isn’t named exactly <code class="language-plaintext highlighter-rouge">Dockerfile</code>.  The last part is the directory
where it will find the <code class="language-plaintext highlighter-rouge">Dockerfile</code> and any files that are referenced
by <code class="language-plaintext highlighter-rouge">COPY</code> (described below).</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>docker build -t training/bwa -f Dockerfile.single-stage .
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="exercise">Exercise</h2>

  <p>Create a <code class="language-plaintext highlighter-rouge">Dockerfile</code> based on this lesson and build it for yourself.</p>

  <blockquote class="solution">
    <h2 id="solution">Solution</h2>

    <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>FROM debian:10-slim
MAINTAINER Peter Amstutz &lt;peter.amstutz@curii.com&gt;

RUN apt-get update -qy
RUN apt-get install -qy build-essential wget unzip zlib1g-dev

# Install BWA 07.7.17
RUN wget https://github.com/lh3/bwa/archive/v0.7.17.zip &amp;&amp; \
	unzip v0.7.17 &amp;&amp; \
	cd bwa-0.7.17 &amp;&amp; \
	make &amp;&amp; \
	cp bwa /usr/bin &amp;&amp; \
	cd .. &amp;&amp; \
	rm -rf bwa-0.7.17
</code></pre></div>    </div>

  </blockquote>
</blockquote>

<h1 id="adding-files-to-the-image-during-the-build">Adding files to the image during the build</h1>

<p>Using the <code class="language-plaintext highlighter-rouge">COPY</code> command, you can copy files from the source directory
(this is the directory your Dockerfile was located) into the image
during the build.  For example, you have a <code class="language-plaintext highlighter-rouge">requirements.txt</code> next to
Dockerfile:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>COPY requirements.txt /tmp/
RUN pip install --requirement /tmp/requirements.txt
</code></pre></div></div>

<h1 id="multi-stage-builds">Multi-stage builds</h1>

<p>As noted, it is good practice to avoiding leaving files in the Docker
image that were required to build the program, but not to run it, as
those files are simply useless bloat.  Docker offers a more
sophisticated way to create clean builds by separating the build steps
from the creation of the final container.  These are called
“multi-stage” builds.</p>

<p>A multi stage build has multiple <code class="language-plaintext highlighter-rouge">FROM</code> lines.  Each <code class="language-plaintext highlighter-rouge">FROM</code> line is a
separate container build.  The last <code class="language-plaintext highlighter-rouge">FROM</code> in the file describes the
final container image that will be created.</p>

<p>The key benefit is that the different stages are independent, but you
can copy files from one stage to another.</p>

<p>Here is an example of the bwa build as a multi-stage build.  It is a
little bit more complicated, but the outcome is a smaller image,
because the “build-essential” tools are not included in the final
image.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Build the base image.  This is the starting point for both the build
# stage and the final stage.
# the "AS base" names the image within the Dockerfile
FROM debian:10-slim AS base
MAINTAINER Peter Amstutz &lt;peter.amstutz@curii.com&gt;

# Install libz, because the bwa binary will depend on it.
# As it happens, this already included in the base Debian distribution
# because lots of things use libz specifically, but it is good practice
# to explicitly declare that we need it.
RUN apt-get update -qy
RUN apt-get install -qy zlib1g


# This is the builder image.  It has the commands to install the
# prerequisites and then build the bwa binary.
FROM base as builder
RUN apt-get install -qy build-essential wget unzip zlib1g-dev

# Install BWA 07.7.17
RUN wget https://github.com/lh3/bwa/archive/v0.7.17.zip
RUN unzip v0.7.17
RUN cd bwa-0.7.17 &amp;&amp; \
    make &amp;&amp; \
    cp bwa /usr/bin


# Build the final image.  It starts from base (where we ensured that
# libz was installed) and then copies the bwa binary from the builder
# image.  The result is the final image only has the compiled bwa
# binary, but not the clutter from build-essentials or from compiling
# the program.
FROM base AS final

# This is the key command, we use the COPY command described earlier,
# but instead of copying from the host, the --from option copies from
# the builder image.
COPY --from=builder /usr/bin/bwa /usr/bin/bwa
</code></pre></div></div>

<h1 id="best-practices-for-docker-images">Best practices for Docker images</h1>

<p><a href="https://docs.docker.com/develop/develop-images/dockerfile_best-practices/">Docker has published guidelines on building efficient images.</a></p>

<p>Some additional considerations when building images for use with Workflows:</p>

<h2 id="store-dockerfiles-in-git-alongside-workflow-definitions">Store Dockerfiles in git, alongside workflow definitions</h2>

<p>Dockerfiles are scripts and should be managed with version control
just like other kinds of code.</p>

<h2 id="be-specific-about-software-versions">Be specific about software versions</h2>

<p>Instead of blindly installing the latest version of a package, or
checking out the <code class="language-plaintext highlighter-rouge">master</code> branch of a git repository and building from
that, be specific in your Dockerfile about what version of the
software you are installing.  This will greatly aid the
reproducibility of your Docker image builds.</p>

<p>Similarly, be as specific as possible about the version of the base
image you want to use in your <code class="language-plaintext highlighter-rouge">FROM</code> command.  If you don’t specify a
tag, the default tag is called “latest”, which can change at any time.</p>

<h2 id="tag-your-builds">Tag your builds</h2>

<p>Use meaningful tags on your own Docker image so you can tell versions
of your Docker image apart as it is updated over time.  These can
reflect the version of the underlying software, or a version you
assign to the Dockerfile itself.  These can be manually assigned
version numbers (e.g. 1.0, 1.1, 1.2, 2.0), timestamps (e.g. YYYYMMDD
like 20220126) or the hash of a git commit.</p>

<h2 id="avoid-putting-reference-data-to-docker-images">Avoid putting reference data to Docker images</h2>

<p>Bioinformatics tools often require large reference data sets to run.
These should be supplied externally (as workflow inputs) rather than
added to the container image.  This makes it easy to update reference
data instead of having to rebuild a new Docker image every time, which
is much more time consuming.</p>

<h2 id="small-scripts-can-be-inputs-too">Small scripts can be inputs, too</h2>

<p>If you have a small script, e.g. a self-contained single-file Python
script which imports Python modules installed inside the container,
you can supply the script as a workflow input.  This makes it easy to
update the script instead of having to rebuild a new Docker image
every time, which is much more time consuming.</p>

<h2 id="dont-use-entrypoint">Don’t use ENTRYPOINT</h2>

<p>The <code class="language-plaintext highlighter-rouge">ENTRYPOINT</code> Dockerfile command modifies the command line that is
executed inside the container.  This can produce confusion when the
command line that supplied to the container and the command that
actually runs are different.</p>

<h2 id="be-careful-about-the-build-cache">Be careful about the build cache</h2>

<p>Docker build has a useful feature where if it has a record of the
exact <code class="language-plaintext highlighter-rouge">RUN</code> command against the exact base layer, it can re-use the
layer from cache instead of re-running it every time.  This is a great
time-saver during development, but can also be a source of
frustration: build steps often download files from the Internet.  If
the file being downloaded changes without the command being used to
download it changing, it will reuse the cached step with the old copy
of the file, instead of re-downloading it.  If this happens, use
<code class="language-plaintext highlighter-rouge">--no-cache</code> to force it to re-run the steps.</p>

<blockquote class="solution">
  <h2 id="episode-solution">Episode solution</h2>
  <ul>
    <li><a href="../assets/answers/ep8/Dockerfile.single-stage">Dockerfile.single-stage</a></li>
    <li><a href="../assets/answers/ep8/Dockerfile.multi-stage">Dockerfile.multi-stage</a></li>
  </ul>
</blockquote>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>Docker images contain the initial state of the filesystem for a container</p>
</li>
    
    <li><p>Docker images are made up of layers</p>
</li>
    
    <li><p>Dockerfiles consist of a series of commands to install software into the container.</p>
</li>
    
  </ul>
</blockquote>

<hr />


</article>

      
      






<footer>
  <div class="row">
    <div class="col-md-6 license" id="license-info" align="left">
	
	Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2018–2022
	by <a href="https://carpentries.org/">The Carpentries</a>
        <br>
        Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2016–2018
	by <a href="https://software-carpentry.org">Software Carpentry Foundation</a>
	
    </div>
    <div class="col-md-6 help-links" align="right">
	
	<a href="/edit//aio.md" data-checker-ignore>Edit on GitHub</a>
	
	/
	<a href="/blob//CONTRIBUTING.md" data-checker-ignore>Contributing</a>
	/
	<a href="/">Source</a>
	/
	<a href="/blob//CITATION" data-checker-ignore>Cite</a>
	/
	<a href="mailto:info@curii.com">Contact</a>
    </div>
  </div>
  <div class="row">
    <div class="col-md-12" align="center">
      Using <a href="https://github.com/carpentries/styles/">The Carpentries style</a>
      version <a href="https://github.com/carpentries/styles/releases/tag/v9.5.3">9.5.3</a>.
    </div>
  </div>
</footer>

      
    </div>
    
<script src="../assets/js/jquery.min.js"></script>
<script src="../assets/js/bootstrap.min.js"></script>
<script src="../assets/js/lesson.js"></script>
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
  ga('create', 'UA-37305346-2', 'auto');
  ga('send', 'pageview');
</script>

  </body>
</html>
