















<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="last-modified" content="2022-04-26 15:35:11 +0000">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- meta "search-domain" used for google site search function google_search() -->
    <meta name="search-domain" value="">
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/lesson.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/syntax.css" />
    <link rel="license" href="#license-info" />
    <script type="text/javascript" src="../assets/js/tabs.js"></script>
    <link rel="stylesheet" href="../assets/css/tabs.css">

    
    <meta http-equiv="refresh" content="0; url=https://doc.arvados.org/rnaseq-cwl-training/02-workflow/index.html" />
    

    



    <!-- Favicons for everyone -->
    <link rel="apple-touch-icon-precomposed" sizes="57x57" href="../assets/favicons/swc/apple-touch-icon-57x57.png" />
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="../assets/favicons/swc/apple-touch-icon-114x114.png" />
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="../assets/favicons/swc/apple-touch-icon-72x72.png" />
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="../assets/favicons/swc/apple-touch-icon-144x144.png" />
    <link rel="apple-touch-icon-precomposed" sizes="60x60" href="../assets/favicons/swc/apple-touch-icon-60x60.png" />
    <link rel="apple-touch-icon-precomposed" sizes="120x120" href="../assets/favicons/swc/apple-touch-icon-120x120.png" />
    <link rel="apple-touch-icon-precomposed" sizes="76x76" href="../assets/favicons/swc/apple-touch-icon-76x76.png" />
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../assets/favicons/swc/apple-touch-icon-152x152.png" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-196x196.png" sizes="196x196" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-96x96.png" sizes="96x96" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-16x16.png" sizes="16x16" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-128.png" sizes="128x128" />
    <meta name="application-name" content="Software Carpentry - Getting started with CWL"/>
    <meta name="msapplication-TileColor" content="#FFFFFF" />
    <meta name="msapplication-TileImage" content="../assets/favicons/swc/mstile-144x144.png" />
    <meta name="msapplication-square70x70logo" content="../assets/favicons/swc/mstile-70x70.png" />
    <meta name="msapplication-square150x150logo" content="../assets/favicons/swc/mstile-150x150.png" />
    <meta name="msapplication-wide310x150logo" content="../assets/favicons/swc/mstile-310x150.png" />
    <meta name="msapplication-square310x310logo" content="../assets/favicons/swc/mstile-310x310.png" />


    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
	<script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
	<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
	<![endif]-->

  <title>
  Create a Workflow by Composing Tools &ndash; Getting started with CWL
  </title>

  </head>
  <body>

    


<div class="panel panel-default life-cycle">
  <div id="life-cycle" class="panel-body beta">
    This lesson is being piloted (Beta version)
  </div>
</div>




    <div class="container">
      
















  
  










<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      
      
      <a href="https://software-carpentry.org" class="pull-left">
        <img class="navbar-logo" src="../assets/img/swc-icon-blue.svg" alt="Software Carpentry logo" />
      </a>
      

      
      <a class="navbar-brand" href="../index.html">Home</a>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">

	
        <li><a href="../CODE_OF_CONDUCT.html">Code of Conduct</a></li>

        
	
        <li><a href="../setup.html">Setup</a></li>

        
        
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Episodes <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
            
            <li><a href="../01-introduction/index.html">Introduction</a></li>
            
            
            <li><a href="../02-workflow/index.html">Create a Workflow by Composing Tools</a></li>
            
            
            <li><a href="../03-running/index.html">Running and Debugging a Workflow</a></li>
            
            
            <li><a href="../04-commandlinetool/index.html">Writing a Tool Wrapper</a></li>
            
            
            <li><a href="../05-scatter/index.html">Analyzing Multiple Samples</a></li>
            
            
            <li><a href="../06-expressions/index.html">Dynamic Workflow Behavior</a></li>
            
            
            <li><a href="../07-resources/index.html"> Resources for further learning</a></li>
            
            
            <li><a href="../08-supplement-docker/index.html">Supplement: Creating Docker Images for Workflows</a></li>
            
	    <li role="separator" class="divider"></li>
            <li><a href="../aio/index.html">All in one page (Beta)</a></li>
          </ul>
        </li>
        
	

	
	
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Extras <span class="caret"></span></a>
          <ul class="dropdown-menu">
            <li><a href="../reference.html">Reference</a></li>
            
            
            <li><a href="../about/index.html">About</a></li>
            
            
            <li><a href="../discuss/index.html">Discussion</a></li>
            
            
            <li><a href="../figures/index.html">Figures</a></li>
            
            
            <li><a href="../guide/index.html">Instructor Notes</a></li>
            
          </ul>
        </li>
	

	
        <li><a href="../LICENSE.html">License</a></li>
	
	<li><a href="/edit//_episodes/02-workflow.md" data-checker-ignore>Improve this page <span class="glyphicon glyphicon-pencil" aria-hidden="true"></span></a></li>
	
      </ul>
      <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form>
    </div>
  </div>
</nav>




























  
  











<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../01-introduction/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
    <h3 class="maintitle"><a href="../">Getting started with CWL</a></h3>
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../03-running/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>

<article>
<div class="row">
  <div class="col-md-1">
  </div>
  <div class="col-md-10">
    <h1 class="maintitle">Create a Workflow by Composing Tools</h1>
  </div>
  <div class="col-md-1">
  </div>
</div>












<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 30 min
      <br/>
      <strong>Exercises:</strong> 10 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>What is the syntax of CWL?</p>
</li>
	
	<li><p>What are the key components of a workflow?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Write a workflow based on the source shell script, making use of existing tool wrappers.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<p>This lesson has migrated to https://doc.arvados.org/rnaseq-cwl-training/02-workflow/index.html</p>

<h1 id="source-shell-script">Source shell script</h1>

<p>In this lesson, we will develop an initial workflow inspired by the
following shell script.</p>

<p>rnaseq_analysis_on_input_file.sh</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">#!/bin/bash</span>

<span class="c"># Based on</span>
<span class="c"># https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/07_automating_workflow.html</span>
<span class="c">#</span>

<span class="c"># This script takes a fastq file of RNA-Seq data, runs FastQC and outputs a counts file for it.</span>
<span class="c"># USAGE: sh rnaseq_analysis_on_input_file.sh &lt;name of fastq file&gt;</span>

<span class="nb">set</span> <span class="nt">-e</span>

<span class="c"># initialize a variable with an intuitive name to store the name of the input fastq file</span>
<span class="nv">fq</span><span class="o">=</span><span class="nv">$1</span>

<span class="c"># grab base of filename for naming outputs</span>
<span class="nv">base</span><span class="o">=</span><span class="sb">`</span><span class="nb">basename</span> <span class="nv">$fq</span> .subset.fq<span class="sb">`</span>
<span class="nb">echo</span> <span class="s2">"Sample name is </span><span class="nv">$base</span><span class="s2">"</span>

<span class="c"># specify the number of cores to use</span>
<span class="nv">cores</span><span class="o">=</span>4

<span class="c"># directory with genome reference FASTA and index files + name of the gene annotation file</span>
<span class="nv">genome</span><span class="o">=</span>rnaseq/reference_data
<span class="nv">gtf</span><span class="o">=</span>rnaseq/reference_data/chr1-hg19_genes.gtf

<span class="c"># make all of the output directories</span>
<span class="c"># The -p option means mkdir will create the whole path if it</span>
<span class="c"># does not exist and refrain from complaining if it does exist</span>
<span class="nb">mkdir</span> <span class="nt">-p</span> rnaseq/results/fastqc
<span class="nb">mkdir</span> <span class="nt">-p</span> rnaseq/results/STAR
<span class="nb">mkdir</span> <span class="nt">-p</span> rnaseq/results/counts

<span class="c"># set up output filenames and locations</span>
<span class="nv">fastqc_out</span><span class="o">=</span>rnaseq/results/fastqc
<span class="nv">align_out</span><span class="o">=</span>rnaseq/results/STAR/<span class="k">${</span><span class="nv">base</span><span class="k">}</span>_
<span class="nv">counts_input_bam</span><span class="o">=</span>rnaseq/results/STAR/<span class="k">${</span><span class="nv">base</span><span class="k">}</span>_Aligned.sortedByCoord.out.bam
<span class="nv">counts</span><span class="o">=</span>rnaseq/results/counts/<span class="k">${</span><span class="nv">base</span><span class="k">}</span>_featurecounts.txt

<span class="nb">echo</span> <span class="s2">"Processing file </span><span class="nv">$fq</span><span class="s2">"</span>

<span class="c"># Run FastQC and move output to the appropriate folder</span>
fastqc <span class="nv">$fq</span>

<span class="c"># Run STAR</span>
STAR <span class="nt">--runThreadN</span> <span class="nv">$cores</span> <span class="nt">--genomeDir</span> <span class="nv">$genome</span> <span class="nt">--readFilesIn</span> <span class="nv">$fq</span> <span class="nt">--outSAMtype</span> BAM SortedByCoordinate <span class="nt">--outSAMunmapped</span> Within

<span class="c"># Create BAM index</span>
samtools index <span class="nv">$counts_input_bam</span>

<span class="c"># Count mapped reads</span>
featureCounts <span class="nt">-T</span> <span class="nv">$cores</span> <span class="nt">-s</span> 2 <span class="nt">-a</span> <span class="nv">$gtf</span> <span class="nt">-o</span> <span class="nv">$counts</span> <span class="nv">$counts_input_bam</span>
</code></pre></div></div>

<h1 id="cwl-syntax">CWL Syntax</h1>

<p>CWL documents are written using a format called “YAML”.  Here is a crash-course in YAML:</p>

<p>Data fields are written with the name, followed by a colon <code class="language-plaintext highlighter-rouge">:</code>, a space,
and then the value.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">fieldName</span><span class="pi">:</span> <span class="s">value</span>
</code></pre></div></div>

<p>The value is the remaining text to the end of the line.</p>

<p>Special characters in YAML include <code class="language-plaintext highlighter-rouge">:</code>, <code class="language-plaintext highlighter-rouge">{</code>, <code class="language-plaintext highlighter-rouge">}</code> <code class="language-plaintext highlighter-rouge">[</code>, <code class="language-plaintext highlighter-rouge">]</code>, <code class="language-plaintext highlighter-rouge">#</code>, <code class="language-plaintext highlighter-rouge">!</code>
and <code class="language-plaintext highlighter-rouge">%</code>.  If your text begins with any of these characters, you must
surround the string in single or double quotes.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">fieldName</span><span class="pi">:</span> <span class="s2">"</span><span class="s">#quoted-value"</span>
</code></pre></div></div>

<p>You can write multi-line text by putting <code class="language-plaintext highlighter-rouge">|-</code> and writing an indented
block.  The leading whitespace will be removed from the actual value.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">fieldName</span><span class="pi">:</span> <span class="pi">|-</span>
  <span class="s">This is a multi-</span>
  <span class="s">line string.</span>
  <span class="s">Horray!</span>
</code></pre></div></div>

<p>Nested sections are indented:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">section1</span><span class="pi">:</span>
  <span class="na">field1</span><span class="pi">:</span> <span class="s">value1</span>
  <span class="na">field2</span><span class="pi">:</span> <span class="s">value2</span>
</code></pre></div></div>

<p>Nested sections can <em>also</em> be wrapped in curly brackets.  In this case, fields must be comma-separated.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">section1</span><span class="pi">:</span> <span class="pi">{</span><span class="nv">field1</span><span class="pi">:</span> <span class="nv">value1</span><span class="pi">,</span> <span class="nv">field2</span><span class="pi">,</span> <span class="nv">value2</span><span class="pi">}</span>
</code></pre></div></div>

<p>When each item is on its own line starting with a dash <code class="language-plaintext highlighter-rouge">-</code>, it is a list.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">section2</span><span class="pi">:</span>
  <span class="pi">-</span> <span class="s">value1</span>
  <span class="pi">-</span> <span class="s">value2</span>
</code></pre></div></div>

<p>List can <em>also</em> be wrapped in square brackets.  In this case, values must be comma-separated.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">section2</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">value1</span><span class="pi">,</span> <span class="nv">value2</span><span class="pi">]</span>
</code></pre></div></div>

<p>Comments start with <code class="language-plaintext highlighter-rouge">#</code>.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># This is a comment about field3</span>
<span class="na">field3</span><span class="pi">:</span> <span class="s">stuff</span>

<span class="na">field4</span><span class="pi">:</span> <span class="s">stuff</span> <span class="c1"># This is a comment about field4</span>
</code></pre></div></div>

<p>Finally, YAML is a superset of JSON.  Valid JSON is also valid YAML,
so you may sometimes see JSON format being used instead of YAML format
for CWL documents.</p>

<h1 id="workflow-header">Workflow header</h1>

<p>Create a new file “main.cwl”</p>

<p>Let’s start with the header.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">cwlVersion</span><span class="pi">:</span> <span class="s">v1.2</span>
<span class="na">class</span><span class="pi">:</span> <span class="s">Workflow</span>
<span class="na">label</span><span class="pi">:</span> <span class="s">RNAseq CWL practice workflow</span>
</code></pre></div></div>

<ul>
  <li>cwlVersion - Every file must include this.  It declares the version of CWL in use.</li>
  <li>class - This is the type of CWL document.  We will see other types in future lessons.</li>
  <li>label - Optional title of your workflow.</li>
</ul>

<h1 id="workflow-inputs">Workflow Inputs</h1>

<p>The purpose of a workflow is to consume some input parameters, run a
series of steps, and produce output values.</p>

<p>For this analysis, the input parameters are the fastq file and the reference data required by STAR.</p>

<p>In the source shell script, the following variables are declared:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># initialize a variable with an intuitive name to store the name of the input fastq file</span>
<span class="nv">fq</span><span class="o">=</span><span class="nv">$1</span>

<span class="c"># directory with genome reference FASTA and index files + name of the gene annotation file</span>
<span class="nv">genome</span><span class="o">=</span>rnaseq/reference_data
<span class="nv">gtf</span><span class="o">=</span>rnaseq/reference_data/chr1-hg19_genes.gtf
</code></pre></div></div>

<p>In CWL, we will declare these variables in the <code class="language-plaintext highlighter-rouge">inputs</code> section.</p>

<p>The inputs section lists each input parameter and its type.  Valid
types include <code class="language-plaintext highlighter-rouge">File</code>, <code class="language-plaintext highlighter-rouge">Directory</code>, <code class="language-plaintext highlighter-rouge">string</code>, <code class="language-plaintext highlighter-rouge">boolean</code>, <code class="language-plaintext highlighter-rouge">int</code>, and
<code class="language-plaintext highlighter-rouge">float</code>.</p>

<p>In this case, the fastq and gene annotation file are individual files.  The STAR index is a directory.  We can describe these inputs in CWL like this:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">inputs</span><span class="pi">:</span>
  <span class="na">fq</span><span class="pi">:</span> <span class="s">File</span>
  <span class="na">genome</span><span class="pi">:</span> <span class="s">Directory</span>
  <span class="na">gtf</span><span class="pi">:</span> <span class="s">File</span>
</code></pre></div></div>

<h1 id="workflow-steps">Workflow Steps</h1>

<p>A workflow consists of one or more steps.  This is the <code class="language-plaintext highlighter-rouge">steps</code> section.</p>

<p>Now we need to describe the first step of the workflow.  In the source
script, the first step is to run <code class="language-plaintext highlighter-rouge">fastqc</code>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Run FastQC and move output to the appropriate folder</span>
fastqc <span class="nv">$fq</span>
</code></pre></div></div>

<p>A workflow step consists of the name of the step, the tool to <code class="language-plaintext highlighter-rouge">run</code>,
the input parameters to be passed to the tool in <code class="language-plaintext highlighter-rouge">in</code>, and the output
parameters expected from the tool in <code class="language-plaintext highlighter-rouge">out</code>.</p>

<p>The value of <code class="language-plaintext highlighter-rouge">run</code> references the tool file.  The tool file describes
how to run the tool (we will discuss how to write tool files in lesson
4).  If we look in <code class="language-plaintext highlighter-rouge">bio-cwl-tools</code> (which you should have imported
when setting up a practice repository in the initial setup
instructions) we find <code class="language-plaintext highlighter-rouge">bio-cwl-tools/fastqc/fastqc_2.cwl</code>.</p>

<p>Next, the <code class="language-plaintext highlighter-rouge">in</code> block is mapping of input parameters to the tool and
the workflow parameters that will be assigned to those inputs.  We
need to know what input parameters the tool accepts.</p>

<p>Let’s open up the tool file and take a look:</p>

<p>Find the <code class="language-plaintext highlighter-rouge">inputs</code> section of <code class="language-plaintext highlighter-rouge">bio-cwl-tools/fastqc/fastqc_2.cwl</code>:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">inputs</span><span class="pi">:</span>

  <span class="na">reads_file</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span>
      <span class="pi">-</span> <span class="s">File</span>
    <span class="na">inputBinding</span><span class="pi">:</span>
      <span class="na">position</span><span class="pi">:</span> <span class="m">50</span>
    <span class="na">doc</span><span class="pi">:</span> <span class="pi">|</span>
      <span class="s">Input bam,sam,bam_mapped,sam_mapped or fastq file</span>
</code></pre></div></div>

<p>Now we know we need to provide an input parameter called <code class="language-plaintext highlighter-rouge">reads_file</code>.</p>

<p>Next, the <code class="language-plaintext highlighter-rouge">out</code> section is a list of output parameters from the tool
that will be used later in the workflow, or as workflow output.  We
need to know what output parameters the tool produces.  Find the
<code class="language-plaintext highlighter-rouge">outputs</code> section of <code class="language-plaintext highlighter-rouge">bio-cwl-tools/fastqc/fastqc_2.cwl</code>:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">outputs</span><span class="pi">:</span>

  <span class="na">zipped_file</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span>
      <span class="pi">-</span> <span class="s">File</span>
    <span class="na">outputBinding</span><span class="pi">:</span>
      <span class="na">glob</span><span class="pi">:</span> <span class="s1">'</span><span class="s">*.zip'</span>
  <span class="na">html_file</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span>
      <span class="pi">-</span> <span class="s">File</span>
    <span class="na">outputBinding</span><span class="pi">:</span>
      <span class="na">glob</span><span class="pi">:</span> <span class="s1">'</span><span class="s">*.html'</span>
  <span class="na">summary_file</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span>
      <span class="pi">-</span> <span class="s">File</span>
    <span class="na">outputBinding</span><span class="pi">:</span>
      <span class="na">glob</span><span class="pi">:</span> <span class="pi">|</span>
        <span class="s">${</span>
          <span class="s">return "*/summary.txt";</span>
        <span class="s">}</span>
</code></pre></div></div>

<p>Now we know to expect an output parameter called <code class="language-plaintext highlighter-rouge">html_file</code>.</p>

<p>Putting this all together, the <code class="language-plaintext highlighter-rouge">fastq</code> step consists of a <code class="language-plaintext highlighter-rouge">run</code>, <code class="language-plaintext highlighter-rouge">in</code>
and <code class="language-plaintext highlighter-rouge">out</code> subsections, and looks like this:</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">steps</span><span class="pi">:</span>
  <span class="na">fastqc</span><span class="pi">:</span>
    <span class="na">run</span><span class="pi">:</span> <span class="s">bio-cwl-tools/fastqc/fastqc_2.cwl</span>
    <span class="na">in</span><span class="pi">:</span>
      <span class="na">reads_file</span><span class="pi">:</span> <span class="s">fq</span>
    <span class="na">out</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">html_file</span><span class="pi">]</span>
</code></pre></div></div>

<h1 id="running-alignment-with-star">Running alignment with STAR</h1>

<p>The next step is to run the STAR aligner.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Run STAR</span>
STAR <span class="nt">--runThreadN</span> <span class="nv">$cores</span> <span class="nt">--genomeDir</span> <span class="nv">$genome</span> <span class="nt">--readFilesIn</span> <span class="nv">$fq</span> <span class="nt">--outSAMtype</span> BAM SortedByCoordinate <span class="nt">--outSAMunmapped</span> Within
</code></pre></div></div>

<p>We will go through the same process as the first section.  We find
there is <code class="language-plaintext highlighter-rouge">bio-cwl-tools/STAR/STAR-Align.cwl</code>.  We will open the file
and look at the <code class="language-plaintext highlighter-rouge">inputs</code> section to determine what input parameters
correspond to the command line parmeters from our source script.</p>

<blockquote class="challenge">
  <h2 id="exercise">Exercise</h2>

  <p>Look at <code class="language-plaintext highlighter-rouge">STAR-Align.cwl</code> and identify the input parameters that
correspond to the command line arguments used in the source script:
<code class="language-plaintext highlighter-rouge">--runThreadN</code>, <code class="language-plaintext highlighter-rouge">--genomeDir</code>, <code class="language-plaintext highlighter-rouge">--outSAMtype</code>, and
<code class="language-plaintext highlighter-rouge">--outSAMunmapped</code>.  Also identify the name of the output parameter.</p>

  <blockquote class="solution">
    <h2 id="solution">Solution</h2>

    <ul>
      <li><code class="language-plaintext highlighter-rouge">--runThreadN</code> → <code class="language-plaintext highlighter-rouge">RunThreadN</code></li>
      <li><code class="language-plaintext highlighter-rouge">--genomeDir</code> → <code class="language-plaintext highlighter-rouge">GenomeDir</code></li>
      <li><code class="language-plaintext highlighter-rouge">--readFilesIn</code> → <code class="language-plaintext highlighter-rouge">ForwardReads</code></li>
      <li><code class="language-plaintext highlighter-rouge">--outSAMtype</code> → <code class="language-plaintext highlighter-rouge">OutSAMtype</code>, <code class="language-plaintext highlighter-rouge">SortedByCoordinate</code></li>
      <li><code class="language-plaintext highlighter-rouge">--outSAMunmapped</code> → <code class="language-plaintext highlighter-rouge">OutSAMunmapped</code></li>
    </ul>

    <p>output parameter name: <code class="language-plaintext highlighter-rouge">alignment</code></p>
  </blockquote>
</blockquote>

<blockquote class="callout">
  <h2 id="command-line-flags">Command line flags</h2>

  <p>Command line flags generally appear appear in either the <code class="language-plaintext highlighter-rouge">arguments</code>
field, or the <code class="language-plaintext highlighter-rouge">prefix</code> field of the <code class="language-plaintext highlighter-rouge">inputBinding</code> section of an
input parameter declaration.  For example, this section of
<code class="language-plaintext highlighter-rouge">STAR-Align.cwl</code> tells us that the <code class="language-plaintext highlighter-rouge">GenomeDir</code> input parameter
corresponds to the <code class="language-plaintext highlighter-rouge">--genomeDir</code> command line parameter.</p>

  <div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="na">GenomeDir</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">Directory</span>
    <span class="na">inputBinding</span><span class="pi">:</span>
      <span class="na">prefix</span><span class="pi">:</span> <span class="s2">"</span><span class="s">--genomeDir"</span>
</code></pre></div>  </div>
</blockquote>

<blockquote class="callout">
  <h2 id="default-values">Default values</h2>

  <p>Sometimes we want to provide input values to a step without making
them as workflow-level inputs.  We can do this with <code class="language-plaintext highlighter-rouge">{default: N}</code>.
For example:</p>

  <div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code>   <span class="na">in</span><span class="pi">:</span>
     <span class="na">RunThreadN</span><span class="pi">:</span> <span class="pi">{</span><span class="nv">default</span><span class="pi">:</span> <span class="nv">4</span><span class="pi">}</span>
</code></pre></div>  </div>
</blockquote>

<blockquote class="challenge">
  <h2 id="exercise-1">Exercise</h2>

  <p>Using the input and output parameters identified in the previous
exercise, write the <code class="language-plaintext highlighter-rouge">run</code>, <code class="language-plaintext highlighter-rouge">in</code> and <code class="language-plaintext highlighter-rouge">out</code> sections of the STAR step.</p>

  <blockquote class="solution">
    <h2 id="solution-1">Solution</h2>

    <div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="na">STAR</span><span class="pi">:</span>
    <span class="na">run</span><span class="pi">:</span> <span class="s">bio-cwl-tools/STAR/STAR-Align.cwl</span>
    <span class="na">in</span><span class="pi">:</span>
      <span class="na">RunThreadN</span><span class="pi">:</span> <span class="pi">{</span><span class="nv">default</span><span class="pi">:</span> <span class="nv">4</span><span class="pi">}</span>
      <span class="na">GenomeDir</span><span class="pi">:</span> <span class="s">genome</span>
      <span class="na">ForwardReads</span><span class="pi">:</span> <span class="s">fq</span>
      <span class="na">OutSAMtype</span><span class="pi">:</span> <span class="pi">{</span><span class="nv">default</span><span class="pi">:</span> <span class="nv">BAM</span><span class="pi">}</span>
      <span class="na">SortedByCoordinate</span><span class="pi">:</span> <span class="pi">{</span><span class="nv">default</span><span class="pi">:</span> <span class="nv">true</span><span class="pi">}</span>
      <span class="na">OutSAMunmapped</span><span class="pi">:</span> <span class="pi">{</span><span class="nv">default</span><span class="pi">:</span> <span class="nv">Within</span><span class="pi">}</span>
    <span class="na">out</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">alignment</span><span class="pi">]</span>
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<h1 id="running-samtools">Running samtools</h1>

<p>The third step is to generate an index for the aligned BAM.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Create BAM index</span>
samtools index <span class="nv">$counts_input_bam</span>
</code></pre></div></div>

<p>For this step, we need to use the output of a previous step as input
to this step.  We refer the output of a step by with name of the step
(STAR), a slash, and the name of the output parameter (alignment), e.g. <code class="language-plaintext highlighter-rouge">STAR/alignment</code></p>

<p>This creates a dependency between steps.  This means the <code class="language-plaintext highlighter-rouge">samtools</code>
step will not run until the <code class="language-plaintext highlighter-rouge">STAR</code> step has completed successfully.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="na">samtools</span><span class="pi">:</span>
    <span class="na">run</span><span class="pi">:</span> <span class="s">bio-cwl-tools/samtools/samtools_index.cwl</span>
    <span class="na">in</span><span class="pi">:</span>
      <span class="na">bam_sorted</span><span class="pi">:</span> <span class="s">STAR/alignment</span>
    <span class="na">out</span><span class="pi">:</span> <span class="pi">[</span><span class="nv">bam_sorted_indexed</span><span class="pi">]</span>
</code></pre></div></div>

<h1 id="featurecounts">featureCounts</h1>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Count mapped reads</span>
featureCounts <span class="nt">-T</span> <span class="nv">$cores</span> <span class="nt">-s</span> 2 <span class="nt">-a</span> <span class="nv">$gtf</span> <span class="nt">-o</span> <span class="nv">$counts</span> <span class="nv">$counts_input_bam</span>
</code></pre></div></div>

<p>As of this writing, the <code class="language-plaintext highlighter-rouge">subread</code> package that provides
<code class="language-plaintext highlighter-rouge">featureCounts</code> is not available in <code class="language-plaintext highlighter-rouge">bio-cwl-tools</code> (and if it has been
added since then, let’s pretend that it isn’t there.)  We will go over
how to write a CWL wrapper for a command line tool in lesson 4.  For
now, we will leave off the final step.</p>

<h1 id="workflow-outputs">Workflow Outputs</h1>

<p>The last thing to do is declare the workflow outputs in the <code class="language-plaintext highlighter-rouge">outputs</code> section.</p>

<p>For each output, we need to declare the type of output, and what
parameter has the output value.</p>

<p>Output types are the same as input types, valid types include <code class="language-plaintext highlighter-rouge">File</code>,
<code class="language-plaintext highlighter-rouge">Directory</code>, <code class="language-plaintext highlighter-rouge">string</code>, <code class="language-plaintext highlighter-rouge">boolean</code>, <code class="language-plaintext highlighter-rouge">int</code>, and <code class="language-plaintext highlighter-rouge">float</code>.</p>

<p>The <code class="language-plaintext highlighter-rouge">outputSource</code> field refers the a step output in the same way that
the <code class="language-plaintext highlighter-rouge">in</code> block does, the name of the step, a slash, and the name of
the output parameter.</p>

<p>For our final outputs, we want the results from fastqc and the
aligned, sorted and indexed BAM file.</p>

<div class="language-yaml highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="na">outputs</span><span class="pi">:</span>
  <span class="na">qc_html</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">outputSource</span><span class="pi">:</span> <span class="s">fastqc/html_file</span>
  <span class="na">bam_sorted_indexed</span><span class="pi">:</span>
    <span class="na">type</span><span class="pi">:</span> <span class="s">File</span>
    <span class="na">outputSource</span><span class="pi">:</span> <span class="s">samtools/bam_sorted_indexed</span>
</code></pre></div></div>

<blockquote class="solution">
  <h2 id="episode-solution">Episode solution</h2>
  <ul>
    <li><a href="../assets/answers/ep2/main.cwl">main.cwl</a></li>
  </ul>
</blockquote>





<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>CWL documents are written using a syntax called YAML.</p>
</li>
    
    <li><p>The key components of the workflow are: the header, the inputs, the steps, and the outputs.</p>
</li>
    
  </ul>
</blockquote>

</article>



























  
  











<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../01-introduction/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../03-running/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>


      
      






<footer>
  <div class="row">
    <div class="col-md-6 license" id="license-info" align="left">
	
	Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2018–2022
	by <a href="https://carpentries.org/">The Carpentries</a>
        <br>
        Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2016–2018
	by <a href="https://software-carpentry.org">Software Carpentry Foundation</a>
	
    </div>
    <div class="col-md-6 help-links" align="right">
	
	<a href="/edit//_episodes/02-workflow.md" data-checker-ignore>Edit on GitHub</a>
	
	/
	<a href="/blob//CONTRIBUTING.md" data-checker-ignore>Contributing</a>
	/
	<a href="/">Source</a>
	/
	<a href="/blob//CITATION" data-checker-ignore>Cite</a>
	/
	<a href="mailto:info@curii.com">Contact</a>
    </div>
  </div>
  <div class="row">
    <div class="col-md-12" align="center">
      Using <a href="https://github.com/carpentries/styles/">The Carpentries style</a>
      version <a href="https://github.com/carpentries/styles/releases/tag/v9.5.3">9.5.3</a>.
    </div>
  </div>
</footer>

      
    </div>
    
<script src="../assets/js/jquery.min.js"></script>
<script src="../assets/js/bootstrap.min.js"></script>
<script src="../assets/js/lesson.js"></script>
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
  ga('create', 'UA-37305346-2', 'auto');
  ga('send', 'pageview');
</script>

  </body>
</html>
